{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation Detection and Overall Mutation Profile Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This scripts detects the mutation rates in each sampled genomeset and uses these values to estimate the mutation rates of the overall SARS-CoV-2 population"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Import the necessary libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from itertools import product\n",
    "import numpy as np\n",
    "import os\n",
    "import pickle\n",
    "from Bio import Phylo, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from collections import defaultdict\n",
    "from scipy import stats\n",
    "import logomaker\n",
    "from mpl_toolkits.axes_grid1.inset_locator import inset_axes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reconstructed_ancestral_sequences(ancestral_state_file_path):\n",
    "    \"\"\"\n",
    "    Reconstructs ancestral node sequences from the ancestral state file.\n",
    "    \n",
    "    Parameters:\n",
    "        - ancestral_state_file_path (str): The path to the ancestral state file.\n",
    "    \n",
    "    Returns:\n",
    "        - ancestral_sequences(dict): A dictionary containing the reconstructed ancestral sequences, \n",
    "                                     where the keys are the ancestral node ids \n",
    "                                     and the values are the reconstructed sequences.\n",
    "    \"\"\"\n",
    "    # Read the ancestral state file\n",
    "    ancestral_states = pd.read_csv(ancestral_state_file_path, sep='\\t', skiprows=8)\n",
    "\n",
    "    # Concatenate the ancestral states for all the positions for a given ancestral node to reconstruct the sequence for that node\n",
    "    ancestral_sequences = {}\n",
    "    for node, group in ancestral_states.groupby('Node'):\n",
    "        sequence = ''.join(group['State'])\n",
    "        ancestral_sequences[node] = sequence\n",
    "    \n",
    "    return ancestral_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#folder_path=\"../Samples/\"   # Folder path to each directory contatining the sample subdirectories\n",
    "#ancestral_sequences= get_reconstructed_ancestral_sequences(folder_path+\"Sample_1/\"+\"Sample_1.state\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert dictionary values into a list\n",
    "#keys = list(ancestral_sequences.keys())\n",
    "#values = list(ancestral_sequences.values())\n",
    "\n",
    "#print(\"Number of ancestral sequences: \", len(values))\n",
    "#set_values = set(values)\n",
    "#print(\"Unique ancestral sequences: \", len(set_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correct_ancestral_sequences(tree, alignment_file_path, ancestral_sequences, outgroup_id):\n",
    "    \"\"\"\n",
    "    Correct ancestral sequences by introducing gaps.\n",
    "    \n",
    "    Parameters:\n",
    "        - tree (tree obj): The phylogenetic tree object\n",
    "        - alignment_file_path(str): The path to the multiple sequence alignment file\n",
    "        - ancestral_sequences (dict): Dictionary containing the reconstructed ancestral sequences, where the keys are the ancestral node ids and the values are the reconstructed sequences.\n",
    "        - outgroup_id(str): Outgroup Sequence ID\n",
    "    Returns:\n",
    "        - ancestral_sequences(dict): Dictionary containing the corrected reconstructed ancestral sequences\n",
    "    \"\"\"\n",
    "    \n",
    "    # Create a dictionary mapping the sample seq ids to their aligned sequences\n",
    "    sample_nodes_sequences = {record.id: str(record.seq) for record in SeqIO.parse(alignment_file_path, \"fasta\")}\n",
    "\n",
    "    # Create a dictionary mapping ancestral nodes to their children nodes and the corresponding distance\n",
    "    ancestral_nodes_to_children = {clade.name: [(child.name, child.branch_length) for child in clade.clades] \n",
    "                                for clade in tree.find_clades() if clade.clades}\n",
    "\n",
    "    # Get the length of the alignment file\n",
    "    alignment_length = len(next(iter(sample_nodes_sequences.values())))  \n",
    "\n",
    "\n",
    "    def weighted_majority_consensus(children_sequences_dict, distance_dict, alignment_length):\n",
    "        \"\"\"\n",
    "        Compute a consensus using children sequences and their distances to the parent.\n",
    "\n",
    "        Parameters:\n",
    "            - children_sequences_dict(dict): Dictionary mapping sequence id to the corresponding sequences\n",
    "            - distance_dict(dict): Dictionary mapping sequence id to their corresponding distance to the parent node\n",
    "        Returns:\n",
    "            - consensus_seq(str): The resulting consensus sequence\n",
    "        \"\"\"\n",
    "        \n",
    "        consensus = []\n",
    "\n",
    "        # Iterate through each position in the alignment length\n",
    "        for i in range(alignment_length):\n",
    "            # Dictionary to store weighted counts of the bases\n",
    "            column_bases = defaultdict(float)  \n",
    "            # Iterate through each of the children aligned sequence\n",
    "            for seq_id, seq in children_sequences_dict.items():\n",
    "                base = seq[i] # Get the base at the ith position of this sequence\n",
    "                # Children closer to the parent will have higher weights for computing the consensus\n",
    "                column_bases[base] += 1/distance_dict[seq_id]  # Add weight for the base which is inverse of the child's distance to the parent\n",
    "\n",
    "            # Select base with highest weighted count\n",
    "            consensus.append(max(column_bases, key=column_bases.get))\n",
    "\n",
    "        # Compute and return the consensus sequence\n",
    "        consensus_seq = \"\".join(consensus)\n",
    "        return consensus_seq\n",
    "\n",
    "    # Reverse the dictionary to process from leaves to root\n",
    "    ancestral_nodes_to_children = dict(reversed(ancestral_nodes_to_children.items()))\n",
    "\n",
    "    # Iterate through each parent node and its children\n",
    "    for parent_node, children_info in ancestral_nodes_to_children.items():\n",
    "        #print(parent_node, children_info)\n",
    "        # Store the children sequences for computing the consensus\n",
    "        children_sequences_dict = {}\n",
    "        # Store weights for weighted consensus\n",
    "        distance_dict = {}  \n",
    "\n",
    "        for child_node, distance in children_info:\n",
    "            distance_dict[child_node] = distance  \n",
    "\n",
    "            if child_node.startswith(\"Node\"):  # Internal node\n",
    "                children_sequences_dict[child_node] = ancestral_sequences[child_node]\n",
    "                #print(\"1:\",distance, \"\\n1:\", ancestral_sequences[child_node].upper())\n",
    "            else:  # Leaf node\n",
    "                # \n",
    "                if child_node != outgroup_id:\n",
    "                    children_sequences_dict[child_node] = sample_nodes_sequences[child_node]\n",
    "                    #print(\"2:\",distance,\"\\n2:\", sample_nodes_sequences[child_node].upper())\n",
    "            \n",
    "        # Compute weighted consensus\n",
    "        consensus_seq = weighted_majority_consensus(children_sequences_dict, distance_dict, alignment_length)\n",
    "        #print(\"C:\", consensus_seq)\n",
    "        # Add gaps to the ancestral genome based on the consensus genome\n",
    "        initial_ancestral_seq = ancestral_sequences[parent_node]\n",
    "        corrected_ancestral_seq = \"\".join(\n",
    "            consensus_seq[i] if consensus_seq[i] == \"-\" else initial_ancestral_seq[i]\n",
    "            for i in range(len(consensus_seq))\n",
    "        )\n",
    "\n",
    "        # Update the ancestral sequences dictionary\n",
    "        ancestral_sequences[parent_node] = corrected_ancestral_seq\n",
    "        #print(\"F:\", corrected_ancestral_seq)\n",
    "    # Return the corrected ancestral sequences dictionary\n",
    "    return ancestral_sequences\n",
    "\n",
    "#correct_ancestral_sequences(Phylo.read(folder_path+\"Sample_1/Sample_1.treefile\", \"newick\"), folder_path+\"Sample_1/\"+\"SARS-CoV-2_Sample_1_5250+1_msa.fasta\", ancestral_sequences, outgroup_id = \"NC_045512.2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to compute the trinucleotide frequency of a sequence to be used in normalization step while computing the trinucleotide mutation rate\n",
    "def get_kmer_frequency(sequence, k):\n",
    "    \"\"\"\n",
    "    Calculate the frequency of k-mers in a given sequence.\n",
    "\n",
    "    Parameters:\n",
    "        - sequence (str): The input RNA sequence.\n",
    "        - k (int): The length of the k-mer.\n",
    "\n",
    "    Returns:\n",
    "        - kmers_frequency_dict(dict): A dictionary containing all possible k-mers as keys and their frequencies in the input sequence as values.\n",
    "    \"\"\"\n",
    "    # Remove gaps from the sequence if any\n",
    "    sequence = sequence.replace(\"-\", \"\")\n",
    "    \n",
    "    #Total number of possible kmers in the sequence\n",
    "    total_kmers = len(sequence) - k + 1\n",
    "\n",
    "    # Declare and Initialize the dictionary to store the kmer frequency by assigning all possible k-mers with frequency 0\n",
    "    nucleotides = [\"C\", \"T\", \"G\", \"A\"]\n",
    "    kmers_frequency_dict = {''.join(combo): 0 for combo in product(nucleotides, repeat=k)}\n",
    "\n",
    "    # Iterate through the sequence to find all k-mers\n",
    "    for i in range(total_kmers):\n",
    "        kmer = sequence[i:i + k]  # Extract the k-mer\n",
    "        # Check if the k-mer consists only of A, T, G, or C and not any other characters\n",
    "        if set(kmer).issubset(set(nucleotides)):\n",
    "            # Increment the frequency of this k-mer in the kmer_frequency dictionary\n",
    "            kmers_frequency_dict[kmer] += 1  \n",
    "    # Return the kmer dictionary\n",
    "    return kmers_frequency_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_mutations_wrt_ancestral_sequence(ancestor_seq, sample_seq):\n",
    "    \"\"\"\n",
    "    To compute the mutations by comparing each sequence against its nearest reconstructed ancestral sequence\n",
    "\n",
    "    Parameters:\n",
    "        - ancestor_seq (str): The ancestor sequence.\n",
    "        - sample_seq (str): The sample sequence.\n",
    "\n",
    "    Returns:\n",
    "        - mutations: A list of mutations, where each mutation is represented as a list containing the mutation, the trinucleotide context, and the trinucleotide frequency\n",
    "    \"\"\"\n",
    "    # Initialize the list to store the mutations\n",
    "    mutations = []\n",
    "    # Get the trinucleotide mutation frequencies in the ancestor sequence\n",
    "    trinucleotide_mutation_frequency = get_kmer_frequency(ancestor_seq, 3)\n",
    "    # Iterate through each position and bases in the two sequences\n",
    "    for i, (a, b) in enumerate(zip(ancestor_seq, sample_seq)):\n",
    "        # If the base is different between the ancestor and sample sequences \n",
    "        if a != b:\n",
    "            #position of the mutation in the alignment is index + 1\n",
    "            position = i + 1  \n",
    "            # Extract the trinucleotide context of the mutation\n",
    "            trinucleotide = str(ancestor_seq[position-2:position+1]).upper()  \n",
    "            # Base in the ancestor sequence\n",
    "            ref_base = a.upper()\n",
    "            # Base in the sample sequence\n",
    "            sample_base = b.upper()\n",
    "            # We ignore those mutations at the starting or end position of the alignment as their trinucleotide context cannot be captured\n",
    "            # We also ignore those mutations in which the surrounding trinucleotide contains abnormal basses or the mutation itself contains abnormal bases\n",
    "            if (\n",
    "                any(base not in [\"C\", \"T\", \"G\", \"A\"] for base in trinucleotide) or\n",
    "                any(base not in [\"C\", \"T\", \"G\", \"A\"] for base in [ref_base, sample_base]) or\n",
    "                len(trinucleotide) != 3\n",
    "            ):                \n",
    "                continue\n",
    "            else:\n",
    "                # Record the mutation data which includes the mutation, its trinucleotide context and trinuleotide frequency \n",
    "                mutation = f\"{ref_base}{position}{sample_base}\".upper()\n",
    "                mutations.append([mutation, trinucleotide, trinucleotide_mutation_frequency[trinucleotide]])\n",
    "               \n",
    "    return mutations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_ancestral_node(tree, node_name):\n",
    "    \"\"\"\n",
    "    To find the nearest ancestral node to a particular node in the tree\n",
    "\n",
    "    Parameters:\n",
    "        - tree: The phylogenetic tree to search in.\n",
    "        - node_name: The name of the node for which we wish to find the nearest ancestor.\n",
    "\n",
    "    Returns:\n",
    "        - ancestral_node: The ancestral node if found, None otherwise.\n",
    "    \"\"\"\n",
    "    # Search the target node in the tree\n",
    "    target_node = next((clade for clade in tree.find_clades() if str(clade.name).split(\"/\")[0] == node_name), None)\n",
    "    if target_node is None:\n",
    "        print(f\"Node '{node_name}' not found.\")\n",
    "        return None\n",
    "\n",
    "    # Traverse the tree to find the nearest ancestral node to the target node\n",
    "    ancestral_node = None\n",
    "    for clade in tree.find_clades():\n",
    "        if target_node in clade:\n",
    "            ancestral_node = clade\n",
    "            break\n",
    "    \n",
    "    #return nearest ancestral node\n",
    "    return ancestral_node\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mutation_data_for_leaf_nodes(alignment_file, ancestral_sequences, tree):\n",
    "    \"\"\"\n",
    "    To compute the mutations for the leaf nodes of the tree\n",
    "    \n",
    "    Parameters:\n",
    "        - alignment_file (str): Path to the multiple sequence alignment file .\n",
    "        - ancestral_sequences (dict): Dictionary of reconstructed ancestral sequences.\n",
    "        - tree: the Phylogenetic tree object.\n",
    "\n",
    "    Returns:\n",
    "        - mutations (dict): Updated dictionary of mutations.\n",
    "        - ancestral_node_data (dict): Updated dictionary of ancestral node data for each leaf node.\n",
    "    \"\"\"\n",
    "    # Initialize the dictionary to store the mutations and ancestral node data\n",
    "    mutations = {}\n",
    "    ancestral_node_data = {}\n",
    "    # Get the leaf node names from the alignment file    \n",
    "    leaf_nodes = [record.id for record in SeqIO.parse(alignment_file, \"fasta\")]\n",
    "    #As the first leaf node is the outgroup sequence, we need to handle it separately\n",
    "    mutations[leaf_nodes[0]] = []\n",
    "    #Remove the reference sequence from the leaf nodes\n",
    "    leaf_nodes = leaf_nodes[1:]\n",
    "    #Compute the mutations for each leaf node\n",
    "    for leaf_node in leaf_nodes:\n",
    "        #find its nearest ancestral node data\n",
    "        ancestral_node = find_ancestral_node(tree, leaf_node)\n",
    "        #store the ancestral node data for the leaf node\n",
    "        ancestral_node_data[leaf_node] = ancestral_node\n",
    "        #extract the sequence of this leaf node\n",
    "        sequence = next((str(record.seq).upper() for record in SeqIO.parse(alignment_file, \"fasta\") if record.id == leaf_node), None)\n",
    "        #compute the mutations against the nearest ancestral sequence\n",
    "        mutations[leaf_node] = find_mutations_wrt_ancestral_sequence(\n",
    "            ancestral_sequences[str(ancestral_node).split(\"/\")[0]], \n",
    "            sequence)         #bootstrapping during tree construction can modify the names of the nodes in the tree by including the bootstrap results, hence we need to extract only the node name from the tree\n",
    "    #return the mutations and ancestral node data\n",
    "    return mutations, ancestral_node_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### For each of the samples, we access the precomputed MSA files and the precomputed tree file. We use the ancestral states at each ancestral nodes of the tree to reconstruct the ancestral sequences. We then compute the mutations for each genome in each sample genome set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sample_size=5250 # Size of each sample\n",
    "folder_path=\"../Samples/\"   # Folder path to each directory contatining the sample subdirectories\n",
    "outgroup_id = \"NC_045512.2\"  # Outgroup Sequence ID\n",
    "\n",
    "# Iterate through each sample\n",
    "for sample in os.listdir(folder_path):\n",
    "    print(sample)\n",
    "    # designate the path to the MSA file for this sample\n",
    "    alignment_file_path = os.path.join(folder_path, sample, f\"SARS-CoV-2_{sample}_{sample_size}+1_msa.fasta\")\n",
    "    # designate the path to the phylogenetic tree for this sample\n",
    "    tree_path = os.path.join(folder_path, sample, f\"{sample}.treefile\")\n",
    "    # read the tree file\n",
    "    tree = Phylo.read(tree_path, \"newick\")\n",
    "    # we now reconstruct the ancestral sequences of the tree from the ancestral state file\n",
    "    ancestral_sequences=get_reconstructed_ancestral_sequences(os.path.join(folder_path, sample, f\"{sample}.state\"))\n",
    "    # correct the ancestral sequences by adding gaps\n",
    "    ancestral_sequences= correct_ancestral_sequences(tree, alignment_file_path, ancestral_sequences, outgroup_id)\n",
    "\n",
    "    #we now compute the mutations for the leaf nodes of the tree   \n",
    "    mutations, ancestral_node_data=get_mutation_data_for_leaf_nodes(alignment_file_path,ancestral_sequences,tree)\n",
    "\n",
    "    # we extract the mutations for the leaf nodes(sampled genomes) of the tree and remove the outgroup sequence\n",
    "    mutations_in_sampled_genomes = {key: value for key, value in mutations.items() if not key.startswith(\"Node\")}\n",
    "    mutations_in_sampled_genomes.pop(outgroup_id)\n",
    "\n",
    "\n",
    "    # we save these files containing the detected mutations\n",
    "    def save_pickle(data, filepath):\n",
    "        # Ensure the directory exists otherwise create it\n",
    "        dir_path = os.path.dirname(filepath)\n",
    "        if not os.path.exists(dir_path):\n",
    "            os.makedirs(dir_path)\n",
    "        try:\n",
    "            with open(filepath, 'wb') as f:\n",
    "                pickle.dump(data, f)\n",
    "            print(f\"Data successfully saved to {filepath}\")\n",
    "        except Exception as e:\n",
    "            print(f\"Error saving to {filepath}: {e}\")\n",
    "\n",
    "    # Save mutations\n",
    "    save_pickle(mutations_in_sampled_genomes, os.path.join(folder_path, sample, f\"SARS-CoV-2_{sample}_mutations_in_sampled_genomes.pkl\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### In the final step, we compute the single and trinucleotide mutation rate estimates for the entire SARS-CoV-2 population and visualize these values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_trinucleotide_mutation_rates(mutations):\n",
    "    \"\"\"\n",
    "    Computes and returns the trinucleotide mutation rates for a given sampled genomeset.\n",
    "\n",
    "    Parameters:\n",
    "        - mutations (dict): The dictionary containing the mutations data for the given sampled genomeset.\n",
    "\n",
    "    Returns: \n",
    "        - trinucleotide_mutation_rates(list): The trinucleotide mutation rates for the given sampled genomeset\n",
    "    \"\"\"\n",
    "    # Initialize the variables\n",
    "    flank_length=1 # since it is a trinucleotide which involves one bases on either side of the mutation site hence the value is set to 1\n",
    "    nucleotides=[\"C\", \"T\", \"G\", \"A\"]\n",
    "    trinucleotide_list=['CCC', 'CCT', 'CCG', 'CCA', 'TCC', 'TCT', 'TCG', 'TCA', 'GCC', 'GCT', 'GCG', 'GCA', 'ACC', 'ACT', 'ACG', 'ACA', 'CTC', 'CTT', 'CTG', 'CTA', 'TTC', 'TTT', 'TTG', 'TTA', 'GTC', 'GTT', 'GTG', 'GTA', 'ATC', 'ATT', 'ATG', 'ATA', 'CGC', 'CGT', 'CGG', 'CGA', 'TGC', 'TGT', 'TGG', 'TGA', 'GGC', 'GGT', 'GGG', 'GGA', 'AGC', 'AGT', 'AGG', 'AGA', 'CAC', 'CAT', 'CAG', 'CAA', 'TAC', 'TAT', 'TAG', 'TAA', 'GAC', 'GAT', 'GAG', 'GAA', 'AAC', 'AAT', 'AAG', 'AAA']\n",
    "        \n",
    "    # Initialize the dictionary to store the mutation data for all the different mutations that have occured in a particular trinucleotide in a given genome population\n",
    "    combined_mutation_data={}\n",
    "    for trinucleotide in trinucleotide_list:\n",
    "        combined_mutation_data[trinucleotide]={}\n",
    "        for nucleotide in nucleotides:\n",
    "            if(trinucleotide[flank_length]!=nucleotide):\n",
    "                combined_mutation_data[trinucleotide][nucleotide]=[]\n",
    "                \n",
    "    # Store the mutation data for all the different mutations that have occured in a particular trinucleotide in the given genome set\n",
    "    for genome in mutations: # Iterate through each genome\n",
    "        # Iterate through each mutation data in that genome\n",
    "        for mutation_in_genome in mutations[genome]:\n",
    "            # Extract the trinucleotide information for the given mutation\n",
    "            trinucleotide=mutation_in_genome[1]\n",
    "            # Extract the new base after mutation\n",
    "            alternate_base=mutation_in_genome[0][-1]\n",
    "            # Append the mutation data to combined_mutation_data dictionary for the given trinucleotide and alternate base\n",
    "            combined_mutation_data[trinucleotide][alternate_base].append(mutation_in_genome)\n",
    "            \n",
    "    # List to store the computed trinucleotide mutation rates for the given genome set\n",
    "    trinucleotide_mutation_rates=[]\n",
    "\n",
    "    # Each mutation type is associated with 16 trinucleotides which we sublist here\n",
    "    for i in range(0,len(combined_mutation_data),16):\n",
    "        sublist=list(combined_mutation_data.keys())[i:i+16]\n",
    "        # Iterate through the nucleotide bases\n",
    "        for new_base in [\"A\",\"C\",\"G\",\"T\"]:\n",
    "            # We iterate through each trinucleotide in the sublist\n",
    "            for trinucleotide in sublist:\n",
    "                # It is a mutation if the middle base in the trinucleotide is not same as the new base\n",
    "                if(trinucleotide[1]!=new_base):\n",
    "                    # We compute and store the trinucleotide mutation rate for this trinucleotide -> new base\n",
    "                    trinucleotide_mutation_rates.append([trinucleotide, new_base, sum(1/i[2] for i in combined_mutation_data[trinucleotide][new_base])/(len(mutations))])\n",
    "    \n",
    "    # Return the list containing the trinucleotide mutation rates\n",
    "    return trinucleotide_mutation_rates\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(\"Results\", exist_ok=True) # Create the directory to save the results\n",
    "os.makedirs(\"Results/overall_mutation_profile\", exist_ok=True) # Create the directory to save the plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_single_nucleotide_mutation_rates(mutations):\n",
    "    \"\"\" \n",
    "    Compute the single nucleotide mutation rates for a given sampled genome set\n",
    "    \n",
    "    Parameters:\n",
    "        - mutations (dict): The dictionary containing the mutations data for the given sampled genomeset.\n",
    "\n",
    "    Returns: \n",
    "        - single_nucleotide_mutation_rates(dict): The single nucleotide mutation rates for the given sampled genomeset \n",
    "    \"\"\"\n",
    "    # Dictionary to store the single nucleotide mutation rates\n",
    "    single_nucleotide_mutation_rates= {'C->A': 0, 'C->G': 0, 'C->T': 0, 'T->A': 0,'T->C': 0,'T->G': 0,'G->A': 0,'G->C': 0,'G->T': 0,'A->C': 0,'A->G': 0,'A->T': 0}\n",
    "    \n",
    "    # Iterate through each genome in the genome set\n",
    "    for genome in mutations:\n",
    "        # Iterate through each mutation in the genome\n",
    "        for mutation in mutations[genome]:\n",
    "            # Get the mutation type\n",
    "            mutation_type=f\"{mutation[0][0]}->{mutation[0][-1]}\"\n",
    "            # Update the single nucleotide mutation rate value in the single_nucleotide_mutation_rates dictionary\n",
    "            single_nucleotide_mutation_rates[mutation_type]+=1/len(mutations)\n",
    "   \n",
    "    # Return the single_nucleotide_mutation_rates dictionary\n",
    "    return single_nucleotide_mutation_rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def compute_single_and_trinucleotide_mutation_rates_estimates(folder_path):\n",
    "    \"\"\"\n",
    "    Compute the Single and Trinucleotide Mutation Rates Estimates for the SARS-CoV-2 Population.\n",
    "\n",
    "    This function processes mutation data from multiple genome samples to compute the mutation rate\n",
    "    estimates for both single-nucleotide and trinucleotide mutations. It calculates the mean rates,\n",
    "    standard deviations, and margins of error (95% confidence) for the different mutation types.\n",
    "\n",
    "    Parameters:\n",
    "        - folder_path (str): Path to the folder containing subdirectories for each sample.\n",
    "                          Each subdirectory must contain a pickle file with mutation data.\n",
    "\n",
    "    Returns:\n",
    "        - single_nucleotide_mutation_types (list): Single-nucleotide mutation types (e.g., \"A→C\").\n",
    "        - single_nucleotide_mutation_rates_mean (list): Mean mutation rates for single-nucleotide mutations.\n",
    "        - single_nucleotide_mutation_rates_margin_of_errors (list): Margins of error for single-nucleotide mutation rates.\n",
    "        - trinucleotide_mutation_types (list): trinucleotide mutation types (e.g., \"ATC→A\").\n",
    "        - mean_trinucleotide_mutation_rates (list): Mean mutation rates for trinucleotide mutations.\n",
    "        - trinucleotide_mutation_rates_margin_of_error (list): Margins of error for trinucleotide mutation rates.\n",
    "    \"\"\"\n",
    "        \n",
    "    # Dictionaries to record the single nucleotide mutation rates and trinucleotide mutation rates for all the samples\n",
    "    trinucleotide_mutation_rates_for_all_samples={}\n",
    "    single_nucleotide_mutation_rates_for_all_samples={}\n",
    "\n",
    "    # Iterate through each sample in the folder path\n",
    "    for sample in sorted(os.listdir(folder_path), key=lambda x: int(x.split(\"_\")[1])):\n",
    "\n",
    "        # Path to the mutation dictionary file for this sample\n",
    "        mutation_dict_path = os.path.join(folder_path, sample, f\"SARS-CoV-2_{sample}_mutations_in_sampled_genomes.pkl\")\n",
    "\n",
    "        # Load the mutation dictionary \n",
    "        with open(mutation_dict_path, 'rb') as file:\n",
    "            mutations_dict = pickle.load(file)\n",
    "        \n",
    "        # Compute the trinucleotide and single nucleotide mutation rates for this sample\n",
    "        trinucleotide_mutation_rates_for_this_sample=get_trinucleotide_mutation_rates(mutations_dict)                                                       \n",
    "        single_nucleotide_mutation_rates_for_this_sample=get_single_nucleotide_mutation_rates(mutations_dict)\n",
    "        \n",
    "\n",
    "        # Assign the trinucleotide and single nucleotide mutation rates values to the dictionaries that contains the data for all the samples\n",
    "        if(sample==\"Sample_1\"):\n",
    "            for trinucleotide_mutation_rate in trinucleotide_mutation_rates_for_this_sample:\n",
    "                trinucleotide_mutation_rates_for_all_samples[(trinucleotide_mutation_rate[0],trinucleotide_mutation_rate[1])]=[trinucleotide_mutation_rate[2]]\n",
    "            for single_nucleotide_mutation_type in single_nucleotide_mutation_rates_for_this_sample:\n",
    "                single_nucleotide_mutation_rates_for_all_samples[single_nucleotide_mutation_type]=[single_nucleotide_mutation_rates_for_this_sample[single_nucleotide_mutation_type]]\n",
    "\n",
    "        else:\n",
    "            for trinucleotide_mutation_rate in trinucleotide_mutation_rates_for_this_sample:\n",
    "                trinucleotide_mutation_rates_for_all_samples[(trinucleotide_mutation_rate[0],trinucleotide_mutation_rate[1])].append(trinucleotide_mutation_rate[2])\n",
    "            for single_nucleotide_mutation_type in single_nucleotide_mutation_rates_for_this_sample:\n",
    "                single_nucleotide_mutation_rates_for_all_samples[single_nucleotide_mutation_type].append(single_nucleotide_mutation_rates_for_this_sample[single_nucleotide_mutation_type])\n",
    "    \n",
    "    # WE NOW ESTIMATE THE SINGLE AND TRINUCLEOTIDE MUTATION RATES FOR THE ENTIRE SARS-CoV-2 POPULATION USING THE VALUES FROM THE INDIVIDUAL SAMPLES\n",
    "    \n",
    "    # Number of Samples\n",
    "    n_samples = len(os.listdir(folder_path))\n",
    "\n",
    "    # Get the corresponding t-score at 95% confidence\n",
    "    confidence_level = 0.95\n",
    "    t_score = stats.t.ppf((1 + confidence_level) / 2.0, n_samples - 1)\n",
    "\n",
    "    # Compute single-nucleotide mutation rate estimates\n",
    "    single_nucleotide_mutation_types = [mutation_type for mutation_type in single_nucleotide_mutation_rates_for_all_samples]\n",
    "    single_nucleotide_mutation_rates_mean = [np.mean(single_nucleotide_mutation_rates_for_all_samples[mutation_type]) for mutation_type in single_nucleotide_mutation_rates_for_all_samples]\n",
    "    single_nucleotide_mutation_rates_std = [np.std(single_nucleotide_mutation_rates_for_all_samples[mutation_type], ddof=1) for mutation_type in single_nucleotide_mutation_rates_for_all_samples]\n",
    "    single_nucleotide_mutation_rates_margin_of_errors = t_score * (np.array(single_nucleotide_mutation_rates_std) / np.sqrt(n_samples))\n",
    "\n",
    "    # Compute trinucleotide mutation rate estimates\n",
    "    trinucleotide_mutation_types = [i[0]+\"→\"+i[-1] for i in trinucleotide_mutation_rates_for_all_samples]\n",
    "    trinucleotide_mutation_rates_mean = [np.mean(trinucleotide_mutation_rates_for_all_samples[trinucleotide_mutation]) for trinucleotide_mutation in trinucleotide_mutation_rates_for_all_samples]  \n",
    "    trinucleotide_mutation_rates_std = [np.std(trinucleotide_mutation_rates_for_all_samples[trinucleotide_mutation], ddof=1) for trinucleotide_mutation in trinucleotide_mutation_rates_for_all_samples]  \n",
    "    trinucleotide_mutation_rates_margin_of_error = t_score * (np.array(trinucleotide_mutation_rates_std) / np.sqrt(n_samples))\n",
    "    \n",
    "    #print(single_nucleotide_mutation_rates_for_all_samples['C->T'])\n",
    "\n",
    "    return single_nucleotide_mutation_types, single_nucleotide_mutation_rates_mean, single_nucleotide_mutation_rates_margin_of_errors, trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error\n",
    "\n",
    "\n",
    "\n",
    "sample_size=5250\n",
    "folder_path=\"../Samples/\"\n",
    "outgroup_id = \"NC_045512.2\"  \n",
    "folder_path=\"../Samples/\"\n",
    "single_nucleotide_mutation_types, single_nucleotide_mutation_rates_mean, single_nucleotide_mutation_rates_margin_of_errors, trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error  = compute_single_and_trinucleotide_mutation_rates_estimates(folder_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plot_single_nucleotide_mutation_rates_estimates(single_nucleotide_mutation_types, single_nucleotide_mutation_rates_mean, single_nucleotide_mutation_rates_margin_of_errors):\n",
    "    \"\"\"\n",
    "    Plot a Bar Plot to Visualize Single-Nucleotide Mutation Rates Estimates for the SARS-CoV-2 Population.\n",
    "\n",
    "    Parameters:\n",
    "        - single_nucleotide_mutation_types (list): List of mutation types\n",
    "        - single_nucleotide_mutation_rates_mean (list): Mean mutation rates for each mutation type.\n",
    "        - single_nucleotide_mutation_rates_margin_of_errors (list): Margins of error for each mutation rate.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Factor for scaling up (10**e) the mutation rates values to make them easier to visualize\n",
    "    e=2\n",
    "\n",
    "    # Colors for the bars\n",
    "    colors = ['#5fbceb', '#06070a', '#d33c33', '#cacaca', '#accc6f', '#e7cac4', '#a5d4e6', '#5e6164', '#cd7430', '#eae8eb', '#dff488', '#f1dcdb']\n",
    "\n",
    "    # Plotting\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    bars = plt.bar([mutation_type.replace(\"T\",\"U\").replace(\"->\", \"→\") for mutation_type in single_nucleotide_mutation_types],     #Since it is RNA we replace T with U\n",
    "                    [mean * 10**e for mean in single_nucleotide_mutation_rates_mean], yerr=[error * 10**e for error in single_nucleotide_mutation_rates_margin_of_errors],\n",
    "                      capsize=5, color=colors[:len(single_nucleotide_mutation_types)])\n",
    "    # Set Axes labels\n",
    "    plt.xlabel(r'Mutation Types $\\rightarrow$')\n",
    "    plt.ylabel(rf'Mutation Rates ($10^{{-{e}}}$ mutations genome$^{{-1}}$) $\\rightarrow$')\n",
    "    # Set Plot Title\n",
    "    #plt.title('Single Nucleotide Mutation Rates Estimates for SARS-CoV-2 Population', fontweight='bold')\n",
    "    # Adjust y-axis limits\n",
    "    plt.ylim(0, 1.5 * max([mean * 10**e for mean in single_nucleotide_mutation_rates_mean]))  \n",
    "\n",
    "    # Adding labels above each bar\n",
    "    for bar, mean, error in zip(bars, single_nucleotide_mutation_rates_mean, single_nucleotide_mutation_rates_margin_of_errors):\n",
    "        plt.text(\n",
    "            bar.get_x() + bar.get_width() / 2, bar.get_height() + error* 10**e,  \n",
    "            f' {round(mean*10**e, 4)} ± {round(error*10**e,4)}', \n",
    "            ha='center', va='bottom', fontsize=10, rotation=90\n",
    "        )\n",
    "    \n",
    "\n",
    "    # Save the plot\n",
    "    plt.savefig('../Results/overall_mutation_profile/single_nucleotide_mutation_rates_estimates.png', dpi=1000)\n",
    "    print(f\"Plot saved to '../Results/overall_mutation_profile/single_nucleotide_mutation_rates_estimates.png'\")\n",
    "    \n",
    "    # Display the plot\n",
    "    plt.show()\n",
    "\n",
    "plot_single_nucleotide_mutation_rates_estimates(single_nucleotide_mutation_types, single_nucleotide_mutation_rates_mean, single_nucleotide_mutation_rates_margin_of_errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sorting in descending order based on mutation rates\n",
    "sorted_data = sorted(\n",
    "    zip(single_nucleotide_mutation_types, single_nucleotide_mutation_rates_mean),\n",
    "    key=lambda x: x[1],\n",
    "    reverse=True\n",
    ")\n",
    "\n",
    "# Unzipping sorted values\n",
    "sorted_mutation_types, sorted_mutation_rates = zip(*sorted_data)\n",
    "\n",
    "# Printing results\n",
    "print(\"Sorted Mutation Types:\", sorted_mutation_types)\n",
    "print(\"Sorted Mutation Rates:\", sorted_mutation_rates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "def plot_trinucleotide_mutation_rates_estimates(trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error):\n",
    "    \"\"\"\n",
    "    Plot to Visualize Trinucleotide Mutation Rates Estimates for all mutation types for the entire SARS-CoV-2 Population.\n",
    "\n",
    "    Parameters:\n",
    "        - trinucleotide_mutation_types (list): List of trinucleotide mutation types (e.g., 'ACG→T').\n",
    "        - trinucleotide_mutation_rates_mean (list): Mean mutation rates for each trinucleotide mutation.\n",
    "        - trinucleotide_mutation_rates_margin_of_error (list): Margins of error for each mean mutation rate.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Factor for scaling up (10**e) the trinucleotide mutation rates values to make them easier to visualize\n",
    "    e=4\n",
    "\n",
    "    # Prepare subheadings corresponding to each single nucleotide mutation types which relates to the 16 associated trinucleotides\n",
    "    subheadings = [trinucleotide_mutation_types[i][1]+\"->\"+trinucleotide_mutation_types[i][-1] for i in range(0, len(trinucleotide_mutation_types), 16)]\n",
    "\n",
    "    # Color cycle for the bars\n",
    "    colors = ['#5fbceb', '#06070a', '#d33c33', '#cacaca', '#accc6f', '#e7cac4', '#a5d4e6', '#5e6164', '#cd7430', '#eae8eb', '#dff488', '#f1dcdb']\n",
    "    color_cycle = [color for color in colors for _ in range(16)]\n",
    "\n",
    "    # Plotting the bar plot with mutation rates estimates and cycling colors\n",
    "    plt.figure(figsize=(30, 7))\n",
    "    for i in range(len(trinucleotide_mutation_types)):\n",
    "        # Plot the bar\n",
    "        bars=plt.bar(trinucleotide_mutation_types[i].replace(\"T\",\"U\"), trinucleotide_mutation_rates_mean[i]*10**e, color=color_cycle[i], edgecolor=color_cycle[i]) \n",
    "\n",
    "        # Add error bars for confidence intervals\n",
    "        plt.errorbar(trinucleotide_mutation_types[i].replace(\"T\",\"U\"), trinucleotide_mutation_rates_mean[i]*10**e, yerr=trinucleotide_mutation_rates_margin_of_error[i]*10**e, fmt='none',  capsize=2, color=\"black\")\n",
    "\n",
    "        # Add subheading at the top middle of every 16 bars\n",
    "        if (i + 1) % 16 == 8:  # the middle point is the 8th bar in each group\n",
    "            bbox_props = dict(boxstyle=\"round,pad=0.5\", edgecolor=color_cycle[i], facecolor=color_cycle[i], alpha=0.5)\n",
    "            plt.text(i, (max(trinucleotide_mutation_rates_mean) + max(trinucleotide_mutation_rates_margin_of_error))*1.4 *10**e, subheadings[int((i + 1) / 16)],\n",
    "                    ha='center', va='bottom', fontsize=12, color='black', bbox=bbox_props)\n",
    "\n",
    "        # Add value (i.e. trinucleotide mutation rates) at the top of each bar\n",
    "        plt.text(i, (trinucleotide_mutation_rates_mean[i] + trinucleotide_mutation_rates_margin_of_error[i])*10**e, \"  \" + str(round(trinucleotide_mutation_rates_mean[i]*10**e, 4))+ \"±\"+ str(round(trinucleotide_mutation_rates_margin_of_error[i]*10**e,4)), ha='center', va='bottom', fontsize=10, rotation=90)\n",
    "\n",
    "    # Plot formatting\n",
    "    plt.xlabel(r'Trinuclotides $\\rightarrow$', fontsize=12)\n",
    "    plt.ylabel(rf\"Trinucleotide Mutation Rate\" + \"\\n\"+rf\"($10^{{-{e}}}$ mutations trinucleotide$^{{-1}}$ genome$^{{-1}}$) $\\rightarrow$\", fontsize=12)\n",
    "\n",
    "    # Rotate x-axis labels for better visibility\n",
    "    plt.xticks(rotation=90, ha='center', fontsize=10)\n",
    "\n",
    "    # Remove outline at top and right\n",
    "    plt.gca().spines['top'].set_visible(False)\n",
    "    plt.gca().spines['right'].set_visible(False)\n",
    "\n",
    "    # Adjust layout\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Display the plot\n",
    "    plt.show()\n",
    "\n",
    "# Call the function to plot the trinucleotide mutation rate estimates for all the mutation types for the entire SARS-CoV-2 population\n",
    "plot_trinucleotide_mutation_rates_estimates(trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plot_trinucleotide_mutations_for_complementary_mutations(trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error, reference_mutation_type, color1, complementary_mutation_type, color2):\n",
    "    \"\"\" \n",
    "    Plot trinucleotide mutation rates for reference and complementary mutation types \n",
    "    along with the Position Weight Matrix (PWM) computed from trinucleotide mutation rates.\n",
    " \n",
    "    Parameters:\n",
    "        - trinucleotide_mutation_types (list): List of trinucleotide mutation types (e.g., 'ACG→T').\n",
    "        - trinucleotide_mutation_rates_mean (list): Mean mutation rates for each trinucleotide mutation.\n",
    "        - trinucleotide_mutation_rates_margin_of_error (list): Margins of error for each mean mutation rate.\n",
    "        - reference_mutation_type (str): The reference mutation type.\n",
    "        - color1 (str): Color for reference mutation type.\n",
    "        - complementary_mutation_type (str): The complementary mutation type.\n",
    "        - color2 (str): Color for complementary mutation type.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Factor for scaling up (10**e) the trinucleotide mutation rates values to make them easier to visualize\n",
    "    e=4\n",
    "\n",
    "    fig, ax = plt.subplots(2,1 , figsize=(7, 10))\n",
    "\n",
    "    # Store the trinucleotide mutation rates estimates and the MOE for the reference and complementary mutation types in a way that makes it easier to plot\n",
    "    trinucleotide_mutation_rate_estimates_and_MOE={}\n",
    "    for index, trinucleotide_mutation_type in enumerate(trinucleotide_mutation_types):\n",
    "        trinucleotide=trinucleotide_mutation_type.split(\"→\")[0]\n",
    "        alternate_base=trinucleotide_mutation_type.split(\"→\")[1]\n",
    "        if(trinucleotide not in trinucleotide_mutation_rate_estimates_and_MOE):\n",
    "            trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide]={}\n",
    "            trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide][alternate_base]=[trinucleotide_mutation_rates_mean[index], trinucleotide_mutation_rates_margin_of_error[index]]\n",
    "        else:\n",
    "            trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide][alternate_base]=[trinucleotide_mutation_rates_mean[index], trinucleotide_mutation_rates_margin_of_error[index]]\n",
    "\n",
    "    trinucleotide_mutation_rates = {}\n",
    "    for trinucleotide in trinucleotide_mutation_rate_estimates_and_MOE:\n",
    "        if trinucleotide[1] == reference_mutation_type[0]:\n",
    "            for alt_base in trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide]:\n",
    "                if alt_base == reference_mutation_type[-1]:\n",
    "                    trinucleotide_mutation_rates[trinucleotide] = [trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide][alt_base]]\n",
    "    for trinucleotide in trinucleotide_mutation_rate_estimates_and_MOE:\n",
    "        if trinucleotide[1] == complementary_mutation_type[0]:\n",
    "            for alt_base in trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide]:\n",
    "                if alt_base == complementary_mutation_type[-1]:\n",
    "                    trinucleotide_mutation_rates[str(Seq(trinucleotide).reverse_complement())].append(trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide][alt_base])\n",
    "\n",
    "    # Plot the bar plot for the reference mutation type\n",
    "    bars1 = ax[0].bar(\n",
    "        [f\"{polynucleotide}_{str(Seq(polynucleotide).reverse_complement())}\" for polynucleotide in trinucleotide_mutation_rates],\n",
    "        [trinucleotide_mutation_rates[polynucleotide][0][0]*10**e for polynucleotide in trinucleotide_mutation_rates],\n",
    "        color=color1)\n",
    "    \n",
    "    ax[0].errorbar([f\"{polynucleotide}_{str(Seq(polynucleotide).reverse_complement())}\" for polynucleotide in trinucleotide_mutation_rates],\n",
    "                    [trinucleotide_mutation_rates[polynucleotide][0][0]*10**e for polynucleotide in trinucleotide_mutation_rates],\n",
    "                     yerr=[trinucleotide_mutation_rates[polynucleotide][0][1]*10**e for polynucleotide in trinucleotide_mutation_rates], fmt='none',  capsize=2, color=\"black\")\n",
    "\n",
    "    ax[0].set_ylabel(rf\"Trinucleotide Mutation Rate\" + \"\\n\"+rf\"($10^{{-{e}}}$ mutations trinucleotide$^{{-1}}$ genome$^{{-1}}$) $\\rightarrow$\")\n",
    "    ax[0].set_title(reference_mutation_type.replace(\"T\",\"U\"), fontweight='bold', fontsize=18)\n",
    "    ax[0].set_ylim(0, 2 * max(trinucleotide_mutation_rates[polynucleotide][0][0]*10**e for polynucleotide in trinucleotide_mutation_rates))\n",
    "    ax[0].set_xticklabels([f\"{polynucleotide} ({str(Seq(polynucleotide).reverse_complement())})\".replace(\"T\",\"U\") for polynucleotide in trinucleotide_mutation_rates], rotation=270)\n",
    "    \n",
    "    # Plot the bar plot for the complementary mutation type\n",
    "    bars2 = ax[1].bar(\n",
    "        [f\"{polynucleotide}_{str(Seq(polynucleotide).reverse_complement())}\" for polynucleotide in trinucleotide_mutation_rates],\n",
    "        [trinucleotide_mutation_rates[polynucleotide][1][0]*10**e for polynucleotide in trinucleotide_mutation_rates],\n",
    "        color=color2)\n",
    "    \n",
    "    ax[1].errorbar([f\"{polynucleotide}_{str(Seq(polynucleotide).reverse_complement())}\" for polynucleotide in trinucleotide_mutation_rates],\n",
    "                    [trinucleotide_mutation_rates[polynucleotide][1][0]*10**e for polynucleotide in trinucleotide_mutation_rates],\n",
    "                     yerr=[trinucleotide_mutation_rates[polynucleotide][1][1]*10**e for polynucleotide in trinucleotide_mutation_rates], fmt='none',  capsize=2, color=\"black\")\n",
    "\n",
    "    ax[1].set_ylabel(rf\"Trinucleotide Mutation Rate\" + \"\\n\"+rf\"($10^{{-{e}}}$ mutations trinucleotide$^{{-1}}$ genome$^{{-1}}$) $\\rightarrow$\")\n",
    "    ax[1].set_title(complementary_mutation_type.replace(\"T\",\"U\"), fontweight='bold', fontsize=18)\n",
    "    ax[1].set_xlabel('Trinucleotides')\n",
    "    ax[1].set_ylim(0, 2 * max(trinucleotide_mutation_rates[polynucleotide][1][0]*10**e for polynucleotide in trinucleotide_mutation_rates))\n",
    "    ax[1].set_xticklabels([f\"{str(Seq(polynucleotide).reverse_complement())} ({polynucleotide})\".replace(\"T\",\"U\") for polynucleotide in trinucleotide_mutation_rates], rotation=270)\n",
    "\n",
    "    # Add Trinucleotide Mutation rate value with error margin at the top of each bar\n",
    "    def add_value_labels(ax, bars, counts, errors):\n",
    "        for bar, count, error in zip(bars, counts, errors):\n",
    "            height = bar.get_height() + error*10**e  \n",
    "            ax.annotate(f' {round(count*10**e,3)}±{round(error*10**e,3)}',\n",
    "                        xy=(bar.get_x() + bar.get_width() / 2, height),\n",
    "                        xytext=(0, 3), \n",
    "                        textcoords=\"offset points\",\n",
    "                        ha='center', va='bottom', rotation=90, fontsize=9)\n",
    "    add_value_labels(\n",
    "        ax[0], bars1, \n",
    "        [trinucleotide_mutation_rates[polynucleotide][0][0] for polynucleotide in trinucleotide_mutation_rates],\n",
    "        [trinucleotide_mutation_rates[polynucleotide][0][1] for polynucleotide in trinucleotide_mutation_rates]\n",
    "    )\n",
    "    add_value_labels(\n",
    "        ax[1], bars2, \n",
    "        [trinucleotide_mutation_rates[polynucleotide][1][0] for polynucleotide in trinucleotide_mutation_rates],\n",
    "        [trinucleotide_mutation_rates[polynucleotide][1][1] for polynucleotide in trinucleotide_mutation_rates]\n",
    "    )\n",
    "\n",
    "\n",
    "    # Add a horizontal line depicting the average trinucleotide mutation rate\n",
    "    ax[0].axhline(np.mean([bar.get_height() for bar in bars1]), color='grey', linestyle='--', linewidth=1)\n",
    "    ax[1].axhline(np.mean([bar.get_height() for bar in bars2]), color='grey', linestyle='--', linewidth=1)\n",
    "\n",
    "    # Compute the PWM for the reference mutation type\n",
    "    combined_mutation_dict = {\n",
    "        trinucleotide: trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide][reference_mutation_type[-1]]\n",
    "        for trinucleotide in trinucleotide_mutation_rate_estimates_and_MOE if trinucleotide[1] == reference_mutation_type[0]\n",
    "    }\n",
    "    pwm = np.zeros((3, 4))\n",
    "    for trinucleotide, rate in combined_mutation_dict.items():\n",
    "        if rate[0] > 0:\n",
    "            for i, base in enumerate(trinucleotide):\n",
    "                if base == \"A\":\n",
    "                    pwm[i, 0] += rate[0]\n",
    "                elif base == \"C\":\n",
    "                    pwm[i, 1] += rate[0]\n",
    "                elif base == \"G\":\n",
    "                    pwm[i, 2] += rate[0]\n",
    "                elif base == \"T\":\n",
    "                    pwm[i, 3] += rate[0]\n",
    "    normalized_pwm = pwm / pwm.sum(axis=1, keepdims=True)\n",
    "    pwm_df = pd.DataFrame(normalized_pwm, index=[i for i in range(1, 4)], columns=[\"A\", \"C\", \"G\", \"U\"])\n",
    "\n",
    "    # Add the PWM sub-plot for the reference mutation type\n",
    "    inset_ax = inset_axes(ax[0], width=\"20%\", height=\"20%\", loc='upper right', borderpad=0)\n",
    "    logo_reference_strand = logomaker.Logo(pwm_df, ax=inset_ax, shade_below=0.5, fade_below=0.5)\n",
    "   \n",
    "    # Remove x and y ticks from the PWM\n",
    "    logo_reference_strand.ax.set_xticks([])\n",
    "    logo_reference_strand.ax.set_yticks([])\n",
    "\n",
    "\n",
    "\n",
    "    # Repeat the same for computing the PWM for the complementary mutation type\n",
    "    combined_mutation_dict = {\n",
    "        trinucleotide: trinucleotide_mutation_rate_estimates_and_MOE[trinucleotide][complementary_mutation_type[-1]]\n",
    "        for trinucleotide in trinucleotide_mutation_rate_estimates_and_MOE if trinucleotide[1] == complementary_mutation_type[0]\n",
    "    }\n",
    "    pwm = np.zeros((3, 4))\n",
    "    for trinucleotide, rate in combined_mutation_dict.items():\n",
    "        if rate[0] > 0:\n",
    "            for i, base in enumerate(trinucleotide):\n",
    "                if base == \"A\":\n",
    "                    pwm[i, 0] += rate[0]\n",
    "                elif base == \"C\":\n",
    "                    pwm[i, 1] += rate[0]\n",
    "                elif base == \"G\":\n",
    "                    pwm[i, 2] += rate[0]\n",
    "                elif base == \"T\":\n",
    "                    pwm[i, 3] += rate[0]\n",
    "    normalized_pwm = pwm / pwm.sum(axis=1, keepdims=True)\n",
    "    pwm_df = pd.DataFrame(normalized_pwm, index=[i for i in range(1, 4)], columns=[\"A\", \"C\", \"G\", \"U\"])\n",
    "\n",
    "\n",
    "    # Plot PWM for complementary mutation type\n",
    "    inset_ax = inset_axes(ax[1], width=\"20%\", height=\"20%\", loc='upper right', borderpad=0)\n",
    "    logo_opposite_strand = logomaker.Logo(pwm_df, ax=inset_ax, shade_below=0.5, fade_below=0.5)\n",
    "    \n",
    "    # Remove x and y ticks from the PWM subplot\n",
    "    logo_opposite_strand.ax.set_xticks([])\n",
    "    logo_opposite_strand.ax.set_yticks([])\n",
    "    \n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Save the plot\n",
    "    plt.savefig(f'../Results/overall_mutation_profile/trinucleotide_mutation_rates_{reference_mutation_type[0]}→{reference_mutation_type[-1]}_{complementary_mutation_type[0]}→{complementary_mutation_type[-1]}.png', dpi=1000)\n",
    "    print(f\"Plot saved to '../Results/overall_mutation_profile/trinucleotide_mutation_rates_{reference_mutation_type[0]}→{reference_mutation_type[-1]}_{complementary_mutation_type[0]}→{complementary_mutation_type[-1]}.png'\")\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "# Call the function to plot trinucleotide mutation rate estimates for the complementary mutation types\n",
    "plot_trinucleotide_mutations_for_complementary_mutations(trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error,  r\"C$\\rightarrow$T\", '#d33c33', r\"G$\\rightarrow$A\", '#a5d4e6')\n",
    "plot_trinucleotide_mutations_for_complementary_mutations(trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error, r\"A$\\rightarrow$G\", '#dff488', r\"T$\\rightarrow$C\", '#accc6f')\n",
    "plot_trinucleotide_mutations_for_complementary_mutations(trinucleotide_mutation_types, trinucleotide_mutation_rates_mean, trinucleotide_mutation_rates_margin_of_error, r\"G$\\rightarrow$T\", '#cd7430', r\"C$\\rightarrow$A\", '#5fbceb')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mutatioanalysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
