{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Temporal Mutation Profile Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### This script computes the temporal mutation rate estimates for each of the 21 timeframes, along with the dominant variant and the geographic distribution of the sequenced genomes during that period."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Import the necessary libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "from Bio.Seq import Seq\n",
    "from collections import OrderedDict\n",
    "import logomaker\n",
    "import os\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from datetime import datetime\n",
    "from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### In the first step compute the trinucleotide and single nucleotide mutation rates per sample for each of the timeframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_single_and_trinucleotide_mutation_rates_per_sample_per_timerame(folder_path, n_months, sample_size, mutation_types):\n",
    "    \"\"\"Computes and Returns the single and trinucleotide mutation rates per sample, separately for all the timeframes\n",
    "    \n",
    "        Parameters:\n",
    "            - folder_path (str): Contains the path to the subdirectories to each sample\n",
    "            - n_months (int): No of months in each time frame\n",
    "            - sample_size (int): No of Genomes in each sample\n",
    "            - mutation_types (list): A list of all the mutation types whose values are to be computed\n",
    "        Returns:\n",
    "            - single_nucleotide_mutation_rates_per_time_period_for_all_samples (nested dictionary): A dictionary containing the single nucleotide mutation rates per time frame for each of the sample\n",
    "            - trinucleotide_mutation_rates_per_time_period_for_all_samples (nested dictionary): A dictionary containing the trinucleotide mutation rates per time frame for each of the sample\n",
    "            - time_periods (list): A list containing the different timeframes\n",
    "    \"\"\"   \n",
    "\n",
    "    # For storing mutation rates for all the samples across the different timeframes\n",
    "    single_nucleotide_mutation_rates_per_time_period_for_all_samples={}\n",
    "    trinucleotide_mutation_rates_per_time_period_for_all_samples={}\n",
    "\n",
    "    # Iterate through each sample\n",
    "    for sample in sorted(os.listdir(folder_path), key=lambda x: int(x.split(\"_\")[1])):\n",
    "        # Allocate the Sample name to the nested dictionary object\n",
    "        single_nucleotide_mutation_rates_per_time_period_for_all_samples[sample]={}\n",
    "        trinucleotide_mutation_rates_per_time_period_for_all_samples[sample]={}\n",
    "        \n",
    "        # We load the metadata file of the sampled genomes for this sample\n",
    "        sampled_genomes_metadata_df=pd.read_csv(os.path.join(folder_path, sample, f\"SARS-CoV-2_{sample}_{sample_size}+1_metadata.csv\"))\n",
    "\n",
    "        # Load the de novo mutation data for this sample \n",
    "        with open(os.path.join(folder_path, sample, f\"SARS-CoV-2_{sample}_mutations_in_sampled_genomes.pkl\"), 'rb') as file:\n",
    "            mutation_data = pickle.load(file)\n",
    "            \n",
    "        # We extract the different time periods from the metadata and sort them based on the date\n",
    "        time_periods=sorted(list(set(sampled_genomes_metadata_df[f\"{n_months}_Month_Period\"].to_list())))\n",
    "\n",
    "        # We iterate through each time period\n",
    "        for time_period in time_periods:\n",
    "            # Allocate the time frame under this timeframe in the nested dictionary\n",
    "            single_nucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period]={}\n",
    "\n",
    "            # Allocate the different mutation types with counts of 0, this will be used to store the mutation rate values for this timeframe\n",
    "            for mutation_type in mutation_types:\n",
    "                single_nucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period][mutation_type]=0\n",
    "\n",
    "            # Initialize the dictionary to store the trinucleotide mutation rates for the different mutation types\n",
    "            trinucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period]={'CCC': {'T': 0, 'G': 0, 'A': 0}, 'CCT': {'T': 0, 'G': 0, 'A': 0}, 'CCG': {'T': 0, 'G': 0, 'A': 0}, 'CCA': {'T': 0, 'G': 0, 'A': 0}, 'TCC': {'T': 0, 'G': 0, 'A': 0}, 'TCT': {'T': 0, 'G': 0, 'A': 0}, 'TCG': {'T': 0, 'G': 0, 'A': 0}, 'TCA': {'T': 0, 'G': 0, 'A': 0}, 'GCC': {'T': 0, 'G': 0, 'A': 0}, 'GCT': {'T': 0, 'G': 0, 'A': 0}, 'GCG': {'T': 0, 'G': 0, 'A': 0}, 'GCA': {'T': 0, 'G': 0, 'A': 0}, 'ACC': {'T': 0, 'G': 0, 'A': 0}, 'ACT': {'T': 0, 'G': 0, 'A': 0}, 'ACG': {'T': 0, 'G': 0, 'A': 0}, 'ACA': {'T': 0, 'G': 0, 'A': 0}, 'CTC': {'C': 0, 'G': 0, 'A': 0}, 'CTT': {'C': 0, 'G': 0, 'A': 0}, 'CTG': {'C': 0, 'G': 0, 'A': 0}, 'CTA': {'C': 0, 'G': 0, 'A': 0}, 'TTC': {'C': 0, 'G': 0, 'A': 0}, 'TTT': {'C': 0, 'G': 0, 'A': 0}, 'TTG': {'C': 0, 'G': 0, 'A': 0}, 'TTA': {'C': 0, 'G': 0, 'A': 0}, 'GTC': {'C': 0, 'G': 0, 'A': 0}, 'GTT': {'C': 0, 'G': 0, 'A': 0}, 'GTG': {'C': 0, 'G': 0, 'A': 0}, 'GTA': {'C': 0, 'G': 0, 'A': 0}, 'ATC': {'C': 0, 'G': 0, 'A': 0}, 'ATT': {'C': 0, 'G': 0, 'A': 0}, 'ATG': {'C': 0, 'G': 0, 'A': 0}, 'ATA': {'C': 0, 'G': 0, 'A': 0}, 'CGC': {'C': 0, 'T': 0, 'A': 0}, 'CGT': {'C': 0, 'T': 0, 'A': 0}, 'CGG': {'C': 0, 'T': 0, 'A': 0}, 'CGA': {'C': 0, 'T': 0, 'A': 0}, 'TGC': {'C': 0, 'T': 0, 'A': 0}, 'TGT': {'C': 0, 'T': 0, 'A': 0}, 'TGG': {'C': 0, 'T': 0, 'A': 0}, 'TGA': {'C': 0, 'T': 0, 'A': 0}, 'GGC': {'C': 0, 'T': 0, 'A': 0}, 'GGT': {'C': 0, 'T': 0, 'A': 0}, 'GGG': {'C': 0, 'T': 0, 'A': 0}, 'GGA': {'C': 0, 'T': 0, 'A': 0}, 'AGC': {'C': 0, 'T': 0, 'A': 0}, 'AGT': {'C': 0, 'T': 0, 'A': 0}, 'AGG': {'C': 0, 'T': 0, 'A': 0}, 'AGA': {'C': 0, 'T': 0, 'A': 0}, 'CAC': {'C': 0, 'T': 0, 'G': 0}, 'CAT': {'C': 0, 'T': 0, 'G': 0}, 'CAG': {'C': 0, 'T': 0, 'G': 0}, 'CAA': {'C': 0, 'T': 0, 'G': 0}, 'TAC': {'C': 0, 'T': 0, 'G': 0}, 'TAT': {'C': 0, 'T': 0, 'G': 0}, 'TAG': {'C': 0, 'T': 0, 'G': 0}, 'TAA': {'C': 0, 'T': 0, 'G': 0}, 'GAC': {'C': 0, 'T': 0, 'G': 0}, 'GAT': {'C': 0, 'T': 0, 'G': 0}, 'GAG': {'C': 0, 'T': 0, 'G': 0}, 'GAA': {'C': 0, 'T': 0, 'G': 0}, 'AAC': {'C': 0, 'T': 0, 'G': 0}, 'AAT': {'C': 0, 'T': 0, 'G': 0}, 'AAG': {'C': 0, 'T': 0, 'G': 0}, 'AAA': {'C': 0, 'T': 0, 'G': 0}}\n",
    "            \n",
    "            # We extract the genomes in each timeframe\n",
    "            sampled_genomes_df_time_period=sampled_genomes_metadata_df[sampled_genomes_metadata_df[f\"{n_months}_Month_Period\"]==time_period]\n",
    "            \n",
    "            # Iterate through each of the genomes in this timeframe            \n",
    "            for genome in set(sampled_genomes_df_time_period[\"Accession\"]):\n",
    "                # Iterate through each mutation in this genome\n",
    "                for mutation in mutation_data[genome]:\n",
    "                    # Generate the mutation type for this mutation\n",
    "                    mutation_type=mutation[0][0]+\"->\"+mutation[0][-1]\n",
    "                    \n",
    "                    # record mutation rates only for the selected mutation types that are mentioned by the user\n",
    "                    if(mutation_type in mutation_types):\n",
    "                        # update the single nucleotide mutation rate dictionary for this timeframe and sample  \n",
    "                        single_nucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period][mutation_type] += 1/len(sampled_genomes_df_time_period)\n",
    "                    # update the trinucleotide mutation rate dictionary for this timeframe and sample\n",
    "                    trinucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period][mutation[1]][mutation[0][-1]]+=(1/(mutation[-1]*len(sampled_genomes_df_time_period)))\n",
    "\n",
    "    return single_nucleotide_mutation_rates_per_time_period_for_all_samples, trinucleotide_mutation_rates_per_time_period_for_all_samples, time_periods\n",
    "\n",
    "sample_size=5250 # No of Genomes per sample\n",
    "folder_path=\"../Samples/\" # Directory to the different sample sub-directories\n",
    "n_months=3  # No of months per timeframe\n",
    "\n",
    "# Top six dominant Mutation types which we are interested in \n",
    "mutation_types=[\"C->T\", \"G->A\", \"A->G\",  \"T->C\", \"G->T\", \"C->A\"]\n",
    "\n",
    "# Call the function to compute the single and trinucleotide mutation rates per sample, separately for all the timeframes\n",
    "single_nucleotide_mutation_rates_per_time_period_for_all_samples, trinucleotide_mutation_rates_per_time_period_for_all_samples, time_periods= get_single_and_trinucleotide_mutation_rates_per_sample_per_timerame(folder_path, n_months, sample_size, mutation_types)\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the directory to store the temporal mutation rates\n",
    "os.makedirs(\"../Results/temporal_mutation_profile\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Next, we plot the single nucleotide mutation rate estimates, along with the dominant variants and geolocation of the sequenced genomes across the different time frames "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimate_and_plot_single_nucleotide_mutation_rates_across_different_timeframes_alongwith_dominant_variants_and_geographic_locations(path_to_all_genomes_metadata, mutation_types, threshold_percentage):\n",
    "    \"\"\"Compute and plot the single nucleotide mutation rate estimates, along with the dominant variants and geolocation of the sequenced genomes across the different time frames \n",
    "    \n",
    "        Parameters:\n",
    "            - path_to_all_genomes_metadata (str): Path to the metadata file of all the genomes\n",
    "            - mutation_types (list): The mutation types in which we are interested for plotting\n",
    "            - threshold_percentage (int): Minimun threshold percentage to select a variant or geolocation as dominant. Those below the threshold_percentage will be grouped as \"Others\"\n",
    "            - mutation_types (list): A list of all the mutation types whose values are to be computed\n",
    "    \"\"\"   \n",
    "\n",
    "    colors=['#d33c33', '#a5d4e6', '#dff488', '#accc6f', '#cd7430', '#5fbceb']\n",
    "\n",
    "    # Load the metadata file containing information about all the genomes\n",
    "    all_genomes_metadata_df = pd.read_csv(path_to_all_genomes_metadata)\n",
    "\n",
    "    # We extract the different time periods from the metadata and sort them based on the date\n",
    "    time_periods=sorted(list(set(all_genomes_metadata_df[f\"{n_months}_Month_Period\"].to_list())))\n",
    "\n",
    "    # -------------------\n",
    "    # WE COMPUTE THE MUTATION RATE ESTIMATES OF THE DIFFERENT MUTATION TYPES ACROSS THE DIFFERENT TIMEFRAMES\n",
    "    # -------------------\n",
    "    \n",
    "    confidence_level = 0.95 # Confidence level for estimation of mutation rates\n",
    "\n",
    "    # Variables to store the mutation rate estimates per timeframe\n",
    "    mean_mutation_rates_per_mutation = {mutation: [] for mutation in mutation_types}\n",
    "    margin_of_errors = {mutation: [] for mutation in mutation_types}\n",
    "\n",
    "    # Iterate through each timeframe\n",
    "    for time_period in time_periods:\n",
    "        # Iterate through each mutation types of interest\n",
    "        for mutation in mutation_types:\n",
    "            # Estimate the mutation rates        \n",
    "            single_nucleotide_mutation_rates_for_this_mutation_type_and_time_period = [single_nucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period][mutation] for sample in single_nucleotide_mutation_rates_per_time_period_for_all_samples]\n",
    "            mean_mutation_rate = np.mean(single_nucleotide_mutation_rates_for_this_mutation_type_and_time_period)\n",
    "            std_mutation_rate=np.std(single_nucleotide_mutation_rates_for_this_mutation_type_and_time_period, ddof=1)\n",
    "            n_samples=len(single_nucleotide_mutation_rates_for_this_mutation_type_and_time_period)\n",
    "            # t-score for 95% CI\n",
    "            t_score = stats.t.ppf((1 + confidence_level) / 2.0, n_samples - 1) \n",
    "            # Calculate margin of error\n",
    "            margin_of_error = t_score * (np.array(std_mutation_rate) / np.sqrt(n_samples))\n",
    "\n",
    "            # Store the computed values for plotting\n",
    "            mean_mutation_rates_per_mutation[mutation].append(mean_mutation_rate)\n",
    "            margin_of_errors[mutation].append(margin_of_error) \n",
    "\n",
    "\n",
    "\n",
    "    # -------------------\n",
    "    # Subplot 3: WE COMPUTE AND PLOT THE PROPORTION OF THE DOMINANT GEOLOCATION OF COLLECTION OF THE GENOMES SEQUENCED ACROSS THE DIFFERENT TIMEFRAMES\n",
    "    # -------------------\n",
    "    \n",
    "\n",
    "    # Remove state/province information from the geolocation, retaining only the name of the country\n",
    "    def get_country(x):\n",
    "            if pd.isna(x):\n",
    "                return \"Unknown\"\n",
    "            if \":\" in x:\n",
    "                if (x.split(\":\")[0]==\"United Kingdom\"):\n",
    "                    return \"UK\"\n",
    "                else:\n",
    "                    return x.split(\":\")[0]\n",
    "            else:\n",
    "                return x\n",
    "    all_genomes_metadata_df[\"Geo_Location\"] = all_genomes_metadata_df[\"Geo_Location\"].apply(get_country)\n",
    "\n",
    "    \n",
    "    # Calculate geolocation counts\n",
    "    geo_counts = all_genomes_metadata_df.groupby(['3_Month_Period', 'Geo_Location']).size().reset_index(name='Count')\n",
    "\n",
    "    # Calculate proportions for Geo_Location\n",
    "    total_counts_geo = geo_counts.groupby('3_Month_Period')['Count'].transform('sum')\n",
    "    geo_counts['Proportion'] = (geo_counts['Count'] / total_counts_geo) * 100\n",
    "\n",
    "    # Group less frequent Geo_Location as \"Others\"\n",
    "    geo_counts['Geo_Location'] = geo_counts.apply(\n",
    "        lambda row: row['Geo_Location'] if row['Proportion'] >= threshold_percentage else 'Others', axis=1\n",
    "    )\n",
    "\n",
    "    # Recompute the Geolocation proportions\n",
    "    grouped_geo_counts = geo_counts.groupby(['3_Month_Period', 'Geo_Location']).sum(numeric_only=True).reset_index()\n",
    "    total_counts_after_grouping_geo = grouped_geo_counts.groupby('3_Month_Period')['Count'].transform('sum')\n",
    "    grouped_geo_counts['Proportion'] = (grouped_geo_counts['Count'] / total_counts_after_grouping_geo) * 100\n",
    "\n",
    "    # Pivot the data for plotting\n",
    "    plot_data_geo = grouped_geo_counts.pivot(index='3_Month_Period', columns='Geo_Location', values='Proportion').fillna(0)\n",
    "\n",
    "    # Sort the columns based on the total proportions of the different Countries\n",
    "    sorted_columns_geo = plot_data_geo.sum(axis=0).sort_values(ascending=False).index\n",
    "    plot_data_geo = plot_data_geo[sorted_columns_geo]\n",
    "\n",
    "\n",
    "    # Assign unique color to each Geolocation\n",
    "    unique_geo_locations = plot_data_geo.columns\n",
    "    palette_geo = sns.color_palette(\"tab20\", len(unique_geo_locations))\n",
    "    color_map_geo = dict(zip(unique_geo_locations, palette_geo))\n",
    "\n",
    "    # For Others use a different colour\n",
    "    if \"Others\" in unique_geo_locations:\n",
    "        color_map_geo[\"Others\"] = (0.5, 0.6, 0.5)  \n",
    "\n",
    "    # Format the time periods for better readability\n",
    "    time_periods = plot_data_geo.index.tolist()\n",
    "    formatted_time_periods = []\n",
    "\n",
    "    for time_period in time_periods:\n",
    "        # Extract the start and end dates\n",
    "        start_date_str, end_date_str = time_period.split('--')\n",
    "        # Convert the date strings to datetime objects\n",
    "        start_date = datetime.strptime(start_date_str, \"%Y-%m-%d\")\n",
    "        end_date = datetime.strptime(end_date_str, \"%Y-%m-%d\")\n",
    "        # Format the dates to \"Month-YYYY\"\n",
    "        start_date_formatted = start_date.strftime(\"%b-%Y\")\n",
    "        end_date_formatted = end_date.strftime(\"%b-%Y\")\n",
    "        # Combine the formatted dates\n",
    "        formatted_date_range = f\"{start_date_formatted} to \\n{end_date_formatted}\"\n",
    "        formatted_time_periods.append(formatted_date_range)\n",
    "\n",
    "    # Plot the Dominant Geolocation stacked barplot \n",
    "    fig, axs = plt.subplots(3, 1, figsize=(20, 18), gridspec_kw={'height_ratios': [1, 1, 1]})\n",
    "    \n",
    "    bar_width = 0.8\n",
    "\n",
    "    ax_geo = axs[2]\n",
    "\n",
    "    # Compute the x and y positions for the stacked barplots \n",
    "    bottom_values_geo = pd.Series([0] * len(plot_data_geo), index=plot_data_geo.index)\n",
    "    x_positions = np.arange(len(plot_data_geo.index))  # Base x-positions for bars\n",
    "\n",
    "    # Plot the stacked barplot\n",
    "    for geo_location, color in color_map_geo.items():\n",
    "        bars_geo = ax_geo.bar(\n",
    "            x_positions,  \n",
    "            plot_data_geo[geo_location],\n",
    "            color=color,\n",
    "            edgecolor=\"black\",\n",
    "            linewidth=0.01,\n",
    "            width=bar_width,\n",
    "            bottom=bottom_values_geo\n",
    "        )\n",
    "        bottom_values_geo += plot_data_geo[geo_location]\n",
    "\n",
    "        # Add labels at the center of the bar)\n",
    "        for bar in bars_geo:\n",
    "            height = bar.get_height()\n",
    "            x = bar.get_x() + bar.get_width() / 2\n",
    "            y = bar.get_y() + height / 2  # Positioning the label at the center of the bar\n",
    "            if height > 0:  # Only annotate non-zero bars\n",
    "                ax_geo.text(\n",
    "                    x, y,\n",
    "                    f\"{geo_location}\\n{height:.1f}%\",  # Label with category and percentage\n",
    "                    ha='center',\n",
    "                    va='center',\n",
    "                    fontsize=9,\n",
    "                    color='black'\n",
    "                )\n",
    "\n",
    "    ax_geo.set_title('c. Geographic distribution of the sequenced genomes across different time frames', fontsize=14, fontweight=\"bold\", pad=20)\n",
    "    ax_geo.set_xlabel(r'Time Frames $\\rightarrow$')\n",
    "\n",
    "    ax_geo.set_ylabel('Proportion (%)')\n",
    "    ax_geo.set_xticks(x_positions)\n",
    "    ax_geo.set_xticklabels(formatted_time_periods, rotation=60, fontsize=9, ha='center', fontweight='bold')\n",
    "\n",
    "\n",
    "\n",
    "    # -------------------\n",
    "    # Subplot 2: Compute and Plot the Pangolin Bar Plot\n",
    "    # -------------------\n",
    "    \n",
    "    ax_pangolin = axs[1]\n",
    "\n",
    "    # Calculate variant count for Pangolin lineage and Geo_Location\n",
    "    variant_counts = all_genomes_metadata_df.groupby(['3_Month_Period', 'Pangolin']).size().reset_index(name='Count')\n",
    "\n",
    "    # Calculate proportions for Pangolin\n",
    "    total_counts_pangolin = variant_counts.groupby('3_Month_Period')['Count'].transform('sum')\n",
    "    variant_counts['Proportion'] = (variant_counts['Count'] / total_counts_pangolin) * 100\n",
    "\n",
    "    # Group less frequent variants as \"Others\" (for Pangolin)\n",
    "    variant_counts['Pangolin'] = variant_counts.apply(\n",
    "        lambda row: row['Pangolin'] if row['Proportion'] >= threshold_percentage else 'Others', axis=1\n",
    "    )\n",
    "    \n",
    "    # Recompute the variant proportions for Pangolin lineages\n",
    "    grouped_variant_counts = variant_counts.groupby(['3_Month_Period', 'Pangolin']).sum(numeric_only=True).reset_index()\n",
    "    total_counts_after_grouping = grouped_variant_counts.groupby('3_Month_Period')['Count'].transform('sum')\n",
    "    grouped_variant_counts['Proportion'] = (grouped_variant_counts['Count'] / total_counts_after_grouping) * 100\n",
    "    \n",
    "    # Pivot the data for plotting Pangolin\n",
    "    plot_data_pangolin = grouped_variant_counts.pivot(index='3_Month_Period', columns='Pangolin', values='Proportion').fillna(0)\n",
    "    # Sort the data based on the overall more dominant variants\n",
    "    sorted_columns_pangolin = plot_data_pangolin.sum(axis=0).sort_values(ascending=False).index\n",
    "    plot_data_pangolin = plot_data_pangolin[sorted_columns_pangolin]\n",
    "\n",
    "    # Assign Unique colour to each Pangolin\n",
    "    unique_pangolins = plot_data_pangolin.columns\n",
    "    palette_pangolin = sns.color_palette(\"tab20\", len(unique_pangolins))\n",
    "    color_map_pangolin = dict(zip(unique_pangolins, palette_pangolin))\n",
    "\n",
    "    #  \"Others\" is depicted in a different color\n",
    "    if \"Others\" in unique_pangolins:\n",
    "        color_map_pangolin[\"Others\"] = (0.5, 0.6, 0.5)  \n",
    "\n",
    "    # Compute the x and y positions for the stacked barplots \n",
    "    bottom_values_pangolin = pd.Series([0] * len(plot_data_pangolin), index=plot_data_pangolin.index)\n",
    "    x_positions = np.arange(len(plot_data_pangolin.index))  \n",
    "\n",
    "    # Plot the stacked barplot\n",
    "    for pangolin, color in color_map_pangolin.items():\n",
    "        bars_pangolin = ax_pangolin.bar(\n",
    "            x_positions,  # Use numerical positions\n",
    "            plot_data_pangolin[pangolin],\n",
    "            color=color,\n",
    "            edgecolor=\"black\",\n",
    "            linewidth=0.01,\n",
    "            width=bar_width,\n",
    "            bottom=bottom_values_pangolin\n",
    "        )\n",
    "        bottom_values_pangolin += plot_data_pangolin[pangolin]\n",
    "\n",
    "        # Record the bar centers for later use\n",
    "        bar_centers = [bar.get_x() + bar.get_width() / 2 for bar in bars_pangolin]\n",
    "        \n",
    "        # Add percentage labels on top of each bar (at the center of the bar)\n",
    "        for bar in bars_pangolin:\n",
    "            height = bar.get_height()\n",
    "            x = bar.get_x() + bar.get_width() / 2\n",
    "            y = bar.get_y() + height / 2  # Positioning the label at the center of the bar\n",
    "            if height > 0:  # Only annotate non-zero bars\n",
    "                ax_pangolin.text(\n",
    "                    x, y,\n",
    "                    f\"{pangolin}\\n{height:.1f}%\",  # Label with category and percentage\n",
    "                    ha='center',\n",
    "                    va='center',\n",
    "                    fontsize=9,\n",
    "                    color='black'\n",
    "                )\n",
    "\n",
    "    ax_pangolin.set_title('b. Proportion of dominant SARS-CoV-2 variants across different time frames', fontsize=14, fontweight=\"bold\", pad=20)\n",
    "    ax_pangolin.set_ylabel('Proportion (%)')\n",
    "    ax_pangolin.set_xticks([])\n",
    "\n",
    "\n",
    "    # -------------------\n",
    "    # Subplot 1: Line Plot for Mutation Rate Estimates \n",
    "    # -------------------\n",
    "    ax_dot = axs[0]\n",
    "\n",
    "    # Plot the mutation rate estimates across different timeframes\n",
    "    for i, mutation in enumerate(mutation_types):\n",
    "        ax_dot.plot(\n",
    "            bar_centers,  # Use bar centers for x-coordinates to align with bar centers\n",
    "            mean_mutation_rates_per_mutation[mutation],\n",
    "            marker='o',\n",
    "            label=mutation.replace(\"T\", \"U\").replace(\"->\", \"→\"),\n",
    "            color=colors[i]\n",
    "        )\n",
    "        ax_dot.fill_between(\n",
    "            bar_centers,  # Align with bar centers\n",
    "            np.array(mean_mutation_rates_per_mutation[mutation]) - np.array(margin_of_errors[mutation]),\n",
    "            np.array(mean_mutation_rates_per_mutation[mutation]) + np.array(margin_of_errors[mutation]),\n",
    "            alpha=0.2,\n",
    "            color=colors[i]\n",
    "        )\n",
    "        ax_dot.errorbar(\n",
    "            bar_centers,  # Align with bar centers\n",
    "            mean_mutation_rates_per_mutation[mutation],\n",
    "            yerr=margin_of_errors[mutation],\n",
    "            fmt='none',\n",
    "            capsize=2,\n",
    "            color=\"grey\"\n",
    "        )\n",
    "\n",
    "    ax_dot.set_title('a. Variation in Mutation Rate Estimates Across Time Frames', fontsize=14, fontweight='bold', pad=20)\n",
    "    \n",
    "    ax_dot.set_ylabel(r'Mutation Rates (mutations genome$^{{-1}}$) $\\rightarrow$')\n",
    "    ax_dot.set_xticks([])  # No x-ticks on the dot plot\n",
    "    ax_dot.grid(True)\n",
    "\n",
    "    # Add legend to the Dot Plot\n",
    "    ax_dot.legend(title=\"Mutation Types\", loc=\"upper left\")\n",
    "\n",
    "    # Sync X-Limits Across All Subplots\n",
    "    for ax in axs:\n",
    "        ax.set_xlim(ax_pangolin.get_xlim())  # Ensure consistent x-limits\n",
    "\n",
    "\n",
    "    # Adjust horizonatl space between the subplots    \n",
    "    plt.subplots_adjust(hspace=0.3)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    # Save the plot\n",
    "    plt.savefig(f'../Results/temporal_mutation_profile/temporal_mutation_profile.png', dpi=1000)\n",
    "    print(f\"Plot saved to '../Results/temporal_mutation_profile/temporal_mutation_profile.png'\")\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "# Threshold for dominant Variant & Geolocatiom percentage\n",
    "threshold_percentage = 5  # Variants and Geolocation below 5% will be grouped as \"Others\"\n",
    "\n",
    "# Path to the metadata file containing information on all the genomes\n",
    "path_to_all_genomes_metadata='../SARS-CoV-2_genomes_metadata_filtered.csv' \n",
    "\n",
    "# Call the function to estimate and plot the single nucleotide mutation rates across different timeframes along with dominant variants and geographic locations\n",
    "estimate_and_plot_single_nucleotide_mutation_rates_across_different_timeframes_alongwith_dominant_variants_and_geographic_locations(path_to_all_genomes_metadata, mutation_types, threshold_percentage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def estimate_and_plot_trinucleotide_mutation_rate_estimates_per_timeframe(trinucleotide_mutation_rates_per_time_period_for_all_samples, time_periods, n_samples):\n",
    "    \"\"\"\n",
    "    Plot the Timewise Trinucleotide Mutation Rates for the Dominant Mutation types\n",
    "    \n",
    "    Parameters:\n",
    "        - trinucleotide_mutation_rates_per_time_period_for_all_samples(dict) : Dictionary contatining Trinucleotide Mutation Rate information from all Samples accross the different timeframes\n",
    "        - time_periods (list): List of the Different Time Frames\n",
    "        - n_samples (int): No of Samples\n",
    "    \"\"\"\n",
    "\n",
    "    # Scaling factor for better visualizing the values\n",
    "    e=4\n",
    "\n",
    "    # Confidence level for estimating the margin of error \n",
    "    confidence_level = 0.95\n",
    "\n",
    "    #Colors assigned  for each mutation type\n",
    "    colors=['#d33c33', '#a5d4e6', '#dff488', '#accc6f', '#cd7430', '#5fbceb']\n",
    "\n",
    "\n",
    "    # List of all possible nucleotides and trinucleotides\n",
    "    nucleotides=[\"C\", \"T\", \"G\", \"A\"]   \n",
    "    trinucleotide_list=['CCC', 'CCT', 'CCG', 'CCA', 'TCC', 'TCT', 'TCG', 'TCA', 'GCC', 'GCT', 'GCG', 'GCA', 'ACC', 'ACT', 'ACG', 'ACA', 'CTC', 'CTT', 'CTG', 'CTA', 'TTC',\n",
    "                         'TTT', 'TTG', 'TTA', 'GTC', 'GTT', 'GTG', 'GTA', 'ATC', 'ATT', 'ATG', 'ATA', 'CGC', 'CGT', 'CGG', 'CGA', 'TGC', 'TGT', 'TGG', 'TGA', 'GGC', 'GGT',\n",
    "                           'GGG', 'GGA', 'AGC', 'AGT', 'AGG', 'AGA', 'CAC', 'CAT', 'CAG', 'CAA', 'TAC', 'TAT', 'TAG', 'TAA', 'GAC', 'GAT', 'GAG', 'GAA', 'AAC', 'AAT', 'AAG', 'AAA']\n",
    "\n",
    "\n",
    "    # FIRST, WE COMPUTE THE ESTIMATES OF THE TRINUCLEOTIDE MUTATION RATES FOR EACH TIME PERIOD FOR THE GIVEN MUTATION TYPES\n",
    "    \n",
    "    # Here, the rows are the time periods and the columns are the mutation types\n",
    "    fig, ax = plt.subplots(len(time_periods),len(mutation_types), figsize=(25, 40))\n",
    "\n",
    "    # We iterate through each mutation type\n",
    "    for col_count, mutation_type in enumerate(mutation_types):\n",
    "        \n",
    "        # Get the colour assigned to this mutation type\n",
    "        color=colors[col_count]\n",
    "\n",
    "        # We iterate through each time period\n",
    "        for row_count, time_period in enumerate(time_periods):\n",
    "            # We create a dictionary that stores the trinucleotide mutation rate estimates in such a way that it becomes easy to plot\n",
    "            trinucleotide_mutation_rates={}\n",
    "            for trinucleotide in trinucleotide_list:\n",
    "                if(trinucleotide[1]==mutation_type[0]):   \n",
    "                    for alternate_base in set(nucleotides) - {trinucleotide[1]}:\n",
    "                        if(alternate_base==mutation_type[-1]):   \n",
    "                            # Compute trinucleotide mutation rate estimates for this timeframe\n",
    "                            mean_mutation_rate = np.mean([trinucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period][trinucleotide][alternate_base] for sample in trinucleotide_mutation_rates_per_time_period_for_all_samples])\n",
    "                            std_mutation_rate = np.std([trinucleotide_mutation_rates_per_time_period_for_all_samples[sample][time_period][trinucleotide][alternate_base] for sample in trinucleotide_mutation_rates_per_time_period_for_all_samples], ddof=1)\n",
    "                            # Calculate margin of error\n",
    "                            t_score = stats.t.ppf((1 + confidence_level) / 2.0, np.array(n_samples) - 1)  # t-score for 95% CI\n",
    "                            margin_of_error = t_score * (np.array(std_mutation_rate) / np.sqrt(n_samples))\n",
    "                            # Store the values for plotting\n",
    "                            trinucleotide_mutation_rates[trinucleotide]=[mean_mutation_rate, margin_of_error]\n",
    "\n",
    "\n",
    "\n",
    "            # We sort the trinucleotide orders in the barplots based on a desired order\n",
    "            if(col_count%2==0 and row_count==len(time_periods)-1):\n",
    "                trinucleotide_order=[str(Seq(i).reverse_complement())  for i in list(trinucleotide_mutation_rates.keys())]\n",
    "            if(col_count%2==1):\n",
    "                trinucleotide_mutation_rates = OrderedDict((trinucleotide, trinucleotide_mutation_rates[trinucleotide]) for trinucleotide in trinucleotide_order)\n",
    "\n",
    "            # NOW WE PLOT THE BARPLOTS DEPICTING THE TRINUCLEOTIDE MUTATION RATES\n",
    "            trinucleotide_mutation_rates_bar_plot = ax[row_count,col_count].bar([f\"{trinucleotide}_{str(Seq(trinucleotide).reverse_complement())}\" for trinucleotide in trinucleotide_mutation_rates],\n",
    "                            [trinucleotide_mutation_rates[trinucleotide][0]*10**e for trinucleotide in trinucleotide_mutation_rates],\n",
    "                            color=color)\n",
    "            \n",
    "            # Add error bars for confidence intervals\n",
    "            ax[row_count,col_count].errorbar([f\"{trinucleotide}_{str(Seq(trinucleotide).reverse_complement())}\" for trinucleotide in trinucleotide_mutation_rates],\n",
    "                                            [trinucleotide_mutation_rates[trinucleotide][0]*10**e for trinucleotide in trinucleotide_mutation_rates],\n",
    "                                            yerr=[trinucleotide_mutation_rates[trinucleotide][1]*10**e for trinucleotide in trinucleotide_mutation_rates],\n",
    "                                            fmt='none',  capsize=2, color=\"black\")\n",
    "\n",
    "            ax[row_count,col_count].axhline(np.mean([trinucleotide_mutation_rates[trinucleotide][0]*10**e for trinucleotide in trinucleotide_mutation_rates]), color='grey', linestyle='--', linewidth=1)\n",
    "            \n",
    "\n",
    "            # We change the time period format for better understanding\n",
    "            \n",
    "            # Split the date range into start and end dates\n",
    "            start_date_str, end_date_str = time_period.split('--')\n",
    "            # Convert the strings to datetime objects\n",
    "            start_date = datetime.strptime(start_date_str, \"%Y-%m-%d\")\n",
    "            end_date = datetime.strptime(end_date_str, \"%Y-%m-%d\")\n",
    "            # Format the dates to \"Month-YYYY\"\n",
    "            start_date_formatted = start_date.strftime(\"%b-%Y\")\n",
    "            end_date_formatted = end_date.strftime(\"%b-%Y\")\n",
    "            # Combine the formatted dates\n",
    "            formatted_date_range = f\"{start_date_formatted} to \\n {end_date_formatted}\"\n",
    "\n",
    "\n",
    "            #Styling\n",
    "            ax[row_count,0].set_ylabel(formatted_date_range, fontsize=8, fontweight='bold') \n",
    "            ax[0,col_count].set_title( mutation_type.replace(\"T\",\"U\").replace(\"->\", r'$\\rightarrow$'), fontweight='bold') \n",
    "            \n",
    "            # Set y-tick label size\n",
    "            ax[row_count,col_count].tick_params(axis='y', labelsize=5)  \n",
    "\n",
    "            # Extend the space on the top for each subplot\n",
    "            ax[row_count,col_count].set_ylim(0, 2.2*10**e*max(trinucleotide_mutation_rates[trinucleotide][0] for trinucleotide in trinucleotide_mutation_rates)) \n",
    "            \n",
    "            # Display the x-ticks(trinucleotides) for the last row only\n",
    "            ax[row_count,col_count].set_xticklabels([])\n",
    "            if(row_count==len(time_periods)-1):\n",
    "                ax[row_count,col_count].set_xticklabels([f\"{trinucleotide} ({str(Seq(trinucleotide).reverse_complement())})\".replace(\"T\",\"U\") for trinucleotide in trinucleotide_mutation_rates], rotation=270)\n",
    "\n",
    "\n",
    "\n",
    "            #To add labels to the top of each bar\n",
    "            def add_value_labels(ax, bars, counts, errors):\n",
    "                    for bar, count, error in zip(bars, counts, errors):\n",
    "                        height = bar.get_height() + error*10**e  # Adjust height to include the margin of error\n",
    "                        ax.annotate(f'  {round(count*10**e,2)} \\n ± {round(error*10**e,2)}',\n",
    "                                    xy=(bar.get_x() + bar.get_width() / 2, height),\n",
    "                                    xytext=(0, 3), \n",
    "                                    textcoords=\"offset points\",\n",
    "                                    ha='center', va='bottom', rotation=90, fontsize=5)\n",
    "\n",
    "            # Add Trinucleotide Mutation rate value with error margin at the top of each bar in both plots\n",
    "            add_value_labels(\n",
    "                    ax[row_count,col_count], trinucleotide_mutation_rates_bar_plot, \n",
    "                    [trinucleotide_mutation_rates[trinucleotide][0] for trinucleotide in trinucleotide_mutation_rates],\n",
    "                    [trinucleotide_mutation_rates[trinucleotide][1] for trinucleotide in trinucleotide_mutation_rates]\n",
    "            )\n",
    "            \n",
    "            \n",
    "            # NOW, WE COMPUTE AND ADD THE POSITION WEIGH MATRIX FROM THE TRINUCLEOTIDE MUTATION RATES\n",
    "\n",
    "            #Initialize the Variable to store compute Position Weight Matrix(PWM) for the given mutation type\n",
    "            pwm = np.zeros((3, 4))\n",
    "            #Compute the Position Weight Matrix(PWM) for the given mutation type \n",
    "            for trinucleotide in trinucleotide_mutation_rates:\n",
    "                rate=trinucleotide_mutation_rates[trinucleotide][0]\n",
    "                if(rate>0):\n",
    "                    for i in range(0, len(trinucleotide)):\n",
    "                        if trinucleotide[i] == \"A\":\n",
    "                            pwm[i, 0] += rate\n",
    "                        elif trinucleotide[i] == \"C\":\n",
    "                            pwm[i, 1] += rate\n",
    "                        elif trinucleotide[i] == \"G\":\n",
    "                            pwm[i, 2] += rate\n",
    "                        elif trinucleotide[i] == \"T\":\n",
    "                            pwm[i, 3] += rate\n",
    "\n",
    "            #Normalize the PWM bases frequency at each position\n",
    "            normalized_pwm = pwm / pwm.sum(axis=1, keepdims=True)\n",
    "            # Create a DataFrame from the Normalized PWM         \n",
    "            pwm_df = pd.DataFrame(normalized_pwm, index=[pos for pos in range(1, 4)], columns=[\"A\", \"C\", \"G\", \"U\"])\n",
    "\n",
    "            # Add the position weight matrix\n",
    "            inset_ax = inset_axes(ax[row_count,col_count], width=\"15%\", height=\"15%\", loc='upper right', borderpad=0)\n",
    "            logo_opposite_strand = logomaker.Logo(pwm_df, ax=inset_ax, shade_below=0.5, fade_below=0.5)\n",
    "            # Remove x and y ticks\n",
    "            logo_opposite_strand.ax.set_xticks([])\n",
    "            logo_opposite_strand.ax.set_yticks([])\n",
    "\n",
    "        \n",
    "        # Adjust the space between the subplots\n",
    "        fig.subplots_adjust(hspace=.1, wspace=.15)  \n",
    "    \n",
    "    plt.tight_layout()\n",
    "    \n",
    "    # Save the plot\n",
    "    plt.savefig(f'../Results/temporal_mutation_profile/temporal_trinucleotide_mutation_rates.png', dpi=500)\n",
    "    print(f\"Plot saved to '../Results/temporal_mutation_profile/temporal_trinucleotide_mutation_rates.png'\")\n",
    "\n",
    "n_samples = 30\n",
    "# Call the function to estimate and plot the trinucleotide mutation rates across different timeframes\n",
    "estimate_and_plot_trinucleotide_mutation_rate_estimates_per_timeframe(trinucleotide_mutation_rates_per_time_period_for_all_samples, time_periods, n_samples)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mutatioanalysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
