{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single-Cell Transcriptomic Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import scanpy as sc\n",
    "import os\n",
    "import gseapy as gp\n",
    "import numpy as np\n",
    "import matplotlib.patheffects as PathEffects\n",
    "from adjustText import adjust_text\n",
    "import matplotlib.pyplot as plt\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "import decoupler as dc\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib.patches import Patch\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### In the first step we access the read count matricies for each of the replicates and then combine them into a single annotation data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_combined_anndata(folder):\n",
    "    \"\"\"\n",
    "    Each of the count matrices belonging to the different samples are accessed and combined to produce the combined annotation data object. \n",
    "   \n",
    "    Parameters:\n",
    "        - folder (str): Folder path containing read count matrix files from the different samples.\n",
    "\n",
    "    Returns:\n",
    "        - combined_anndata (adata): A combined Annotation Data object generated from count matrices.\n",
    "    \"\"\"\n",
    "\n",
    "    anndata_list=[]\n",
    "    # Iterate through the count matrix files\n",
    "    for filename in os.listdir(folder):\n",
    "        # Print the filename being accessed\n",
    "        print(filename)\n",
    "        # Access the .csv count matrices \n",
    "        if filename.endswith(\".csv\"):\n",
    "            #Load the AnnData object\n",
    "            anndata = sc.read_csv(folder + \"/\" + filename).T        \n",
    "            \n",
    "            #Append the AnnData object to the list\n",
    "            anndata_list.append(anndata)\n",
    "\n",
    "            print(f\" successfully loaded {folder}/{filename}\")\n",
    "\n",
    "    # Generate the combined AnnData object\n",
    "    combined_anndata = sc.concat(anndata_list)\n",
    "    \n",
    "    # Return the combined AnnData object\n",
    "    return combined_anndata\n",
    "\n",
    "# Call the function to generate the combined AnnData object\n",
    "combined_anndata = generate_combined_anndata(\"../Single-Cell-Transcriptomics/GSE171524_RAW\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### We load the medatada file containing the cell information and use it for celltype annotation of the cells in the combined Anndata object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the metadata file\n",
    "metadata=pd.read_csv(\"../Single-Cell-Transcriptomics/GSE171524_lung_metaData.txt\",sep=\"\\t\", index_col=0)\n",
    "\n",
    "# Annotate the annotation data object using the information from the metadata file\n",
    "combined_anndata.obs = combined_anndata.obs.merge(metadata, left_index=True, right_index=True, how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### We now separate the AT2 cells for further investigation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types= [\"AT2\"]\n",
    "\n",
    "#Keep only this cell type and filter our the rest\n",
    "combined_anndata=combined_anndata[combined_anndata.obs[\"cell_type_fine\"].isin(cell_types)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### We check the total number of AT2 Cells in the combined Annotation data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_anndata.obs.groupby(\"cell_type_fine\").count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### We check the cell counts for each biosample used in the study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_anndata.obs.groupby([\"cell_type_fine\", \"biosample_id\"]).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### We check the total number of AT2 Cells in the combined Annotation data object coming from the SARS-CoV-2 patients and Control samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_anndata.obs.groupby(\"disease__ontology_label\").count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Next, we carry out differential expression analysis through pseudobulking and visualize the results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the pseudo-bulk profiles\n",
    "pseudobulk_data = dc.get_pseudobulk(\n",
    "    combined_anndata,\n",
    "    sample_col='biosample_id',\n",
    "    groups_col='cell_type_intermediate',\n",
    "    mode='sum',\n",
    "    min_cells=25,\n",
    "    min_counts=2500\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the pseudobulk replicates\n",
    "pseudobulk_data.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number of infected and control pseudobulk replicates\n",
    "pseudobulk_data.obs.groupby(\"disease__ontology_label\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Carry out DEA using Deseq2\n",
    "dds = DeseqDataSet(\n",
    "    adata=pseudobulk_data,\n",
    "    design_factors=['disease__ontology_label'],\n",
    ")\n",
    "\n",
    "# Generate and get the DEA analysis results\n",
    "dds.deseq2()\n",
    "stat_ressults = DeseqStats(dds, contrast=('disease--ontology-label', 'COVID-19', 'normal'))\n",
    "stat_ressults.summary()\n",
    "dea_results  = stat_ressults.results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The genes of interst we want to investigate\n",
    "antiviral_zap=[\"ZC3HAV1\"]\n",
    "apobecs=['AICDA', 'APOBEC1', 'APOBEC2', 'APOBEC3A', 'APOBEC3B', 'APOBEC3C', 'APOBEC3D', 'APOBEC3F', 'APOBEC3G', 'APOBEC3H', 'APOBEC4']\n",
    "adars=['ADAR', 'ADARB1', 'ADARB2']\n",
    "gene_list=antiviral_zap+apobecs+adars\n",
    "\n",
    "# Make directory to store the single cell analysis images\n",
    "os.makedirs('../Results/sc_transcriptomic_analysis/', exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# See the DESEQ2 results for the genes of interest\n",
    "dea_results[dea_results.index.isin(gene_list)].to_csv(\"../Results/sc_transcriptomic_analysis/DESEQ2_results_genes_of_interest.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the DEA results volcano plots highlighting the genes of interests\n",
    "\n",
    "def plot_volcano(data, gene_list, logfc_threshold=.75, padj_threshold=0.05, adjust=True):\n",
    "    \"\"\"\n",
    "    Plot a volcano plot to visualize differential gene expression analysis results.\n",
    "\n",
    "    Parameters:\n",
    "    - data (pandas.DataFrame): The data containing gene expression analysis results.\n",
    "    - gene_list (list): A list of genes of interest to be highlighted in the plot.\n",
    "    - logfc_threshold (float, optional): The threshold for log2 fold change. Default is 0.75.\n",
    "    - padj_threshold (float, optional): The threshold for adjusted p-value. Default is 0.1.\n",
    "    - adjust (bool, optional): Whether to adjust the position of gene labels to avoid overlap. Default is True.\n",
    "    \"\"\"\n",
    "    plt.figure(figsize=(10, 10))\n",
    "\n",
    "    # Filter points based on thresholds\n",
    "    nonsig_points = data[(data['padj'] > padj_threshold) | (abs(data['log2FoldChange']) < logfc_threshold)]\n",
    "    sig_points = data[(data['padj'] <= padj_threshold) & (abs(data['log2FoldChange']) >= logfc_threshold)]\n",
    "    sig_pval_only = data[data['padj'] <= padj_threshold]\n",
    "\n",
    "    # Plot non-significant points\n",
    "    plt.scatter(nonsig_points['log2FoldChange'], -np.log10(nonsig_points['padj']), c='#DCDCDC', edgecolor='#E0E0E0', label='Not Statistically Significant', marker='o')\n",
    "\n",
    "    # Plot significant points\n",
    "    plt.scatter(sig_points['log2FoldChange'], -np.log10(sig_points['padj']), c='#A9A9A9', edgecolor='#E0E0E0', label='Statistically Significant', marker='o')\n",
    "\n",
    "    # Plot genes in the filtered gene list with a different marker\n",
    "    genes_of_interest = data[data.index.isin(gene_list)]\n",
    "    plt.scatter(genes_of_interest['log2FoldChange'], -np.log10(genes_of_interest['padj']), c='b', edgecolor='none', label='Genes of Interest', marker='^')\n",
    "\n",
    "    # Draw horizontal line for padj threshold\n",
    "    plt.axhline(y=-np.log10(padj_threshold), color='#909090', linestyle='--', linewidth=1)\n",
    "\n",
    "    # Draw vertical lines for log2 fold change thresholds\n",
    "    plt.axvline(x=logfc_threshold, color='#909090', linestyle='--', linewidth=1)\n",
    "    plt.axvline(x=-logfc_threshold, color='#909090', linestyle='--', linewidth=1)\n",
    "\n",
    "    # Label points in the filtered gene list\n",
    "    texts = []\n",
    "    for gene in gene_list:\n",
    "        gene_data = data[data.index == gene]\n",
    "        if not gene_data.empty:\n",
    "            x = gene_data['log2FoldChange'].values[0]\n",
    "            y = -np.log10(gene_data['padj'].values[0])\n",
    "            gene_label = gene + (r'$^{★}$' if gene in sig_pval_only.index else '')\n",
    "            txt = plt.text(x, y, gene_label, size=12, ha='center', va='center')  \n",
    "            txt.set_path_effects([PathEffects.withStroke(linewidth=3, foreground='w')])\n",
    "            texts.append(txt)\n",
    "    \n",
    "    # Add legend for the statistically significant genes of interest which are highlighted with a '★' marker\n",
    "    plt.scatter([], [], c='black', marker=\"*\",edgecolor='none', s=70, label=f'Statistically Significant Genes of \\nInterest ($p_{{adj}}$ < {padj_threshold})')\n",
    "\n",
    "\n",
    "    plt.legend(fontsize=12) \n",
    "    plt.xlabel(\"$log_{2}$ fold change\", size=15)\n",
    "    plt.ylabel(\"-$log_{10}$ $p_{adj}$-value\", size=15)\n",
    "\n",
    "    if adjust:\n",
    "        adjust_text(texts, arrowprops=dict(arrowstyle=\"-\", color='k', lw=0.5))\n",
    "        \n",
    "    # Save the plot\n",
    "    plt.savefig('../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_volcano_plot.png', dpi = 1000)\n",
    "    print(\"The Volcano plot has been saved at '../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_volcano_plot.png'\")\n",
    "    \n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n",
    "# Call the function to plot the volcano plot highlighting the genes of interest\n",
    "plot_volcano(dea_results, gene_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Carry out Geneset Enrichment Analysis of the Oxidative damage reponse geneset\n",
    "\n",
    "def run_gene_set_enrichment_analysis(dea_results, oxidative_stress_gene_set):\n",
    "    \"\"\"\n",
    "    Runs gene set enrichment analysis using GSEApy for a specified gene set.\n",
    "\n",
    "    Parameters:\n",
    "        - dea_results(pandas dataframe): DESeq2 results with columns 'GeneSymbol', 'padj', and 'log2FoldChange'.\n",
    "        - oxidative_stress_gene_set (str): Path to the Oxidative Damage response gene set file.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a ranking of genes based on the DESeq2 results\n",
    "    dea_results[\"-log_padj x lfc\"] = -np.log(dea_results[\"padj\"]) * dea_results[\"log2FoldChange\"]\n",
    "    # Sort the genes by the -log_padj x lfc\n",
    "    ranked_genes = dea_results[['GeneSymbol', '-log_padj x lfc']].dropna().sort_values('-log_padj x lfc', ascending=False)\n",
    "    # Reset the index to GeneSymbol column\n",
    "    ranked_genes.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    # Run Gene Set Enrichment Analysis\n",
    "    pre_res = gp.prerank(rnk=ranked_genes, gene_sets=oxidative_stress_gene_set, seed=6, min_size=0, max_size=200)\n",
    "\n",
    "    # Plot the results\n",
    "    pre_res.plot(terms=\"WP_OXIDATIVE_DAMAGE_RESPONSE\", figsize=(10, 10))\n",
    "    \n",
    "    # Save the figure\n",
    "    plt.savefig(\"../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_gsea_plot.png\", dpi=1000)\n",
    "    print(\"The GSEA plot has been saved at '../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_gsea_plot.png'\")\n",
    "    \n",
    "# Explicitly Add the GeneSymbol column to the DataFrame as it would be needed for GSEA\n",
    "dea_results[\"GeneSymbol\"]=dea_results.index\n",
    "\n",
    "# Call the function to run gene set enrichment analysis\n",
    "run_gene_set_enrichment_analysis(dea_results, \"../WP_OXIDATIVE_DAMAGE_RESPONSE.gmt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the log fold changes heatmap for the genes of interest\n",
    "\n",
    "def plot_lfd_heatmap(dea_results, gene_list):\n",
    "    \"\"\"\n",
    "    Plot the Log2 Fold Changes heatmap for the Genes of Interest\n",
    "    \n",
    "    Parameters:\n",
    "        - dea_results(pandas dataframe): DESeq2 results with columns 'GeneSymbol', 'padj', and 'log2FoldChange'.\n",
    "        - gene_list (list): List of the Gene Symbols of the genes of interest.\n",
    "    \"\"\"\n",
    "    # Extract the Log2 Fold Changes for the genes of interest\n",
    "    log2_fold_changes = [\n",
    "        dea_results.loc[dea_results[\"GeneSymbol\"] == gene, \"log2FoldChange\"].values[0]\n",
    "        if not dea_results.loc[dea_results[\"GeneSymbol\"] == gene, \"log2FoldChange\"].empty\n",
    "        else None\n",
    "        for gene in gene_list\n",
    "    ]\n",
    "\n",
    "    # Create a DataFrame for the log2 fold changes, keeping missing values (None)\n",
    "    log2_fold_changes_df = pd.DataFrame(\n",
    "        data=log2_fold_changes, \n",
    "        index=gene_list, \n",
    "        columns=[\"log2FoldChange\"]\n",
    "    )\n",
    "\n",
    "    # Plotting the heatmap, including missing values as NaN (None will appear as NaN)\n",
    "    plt.figure(figsize=(0.75, 11))  # Dynamically adjust height\n",
    "    heatmap = sns.heatmap(\n",
    "        log2_fold_changes_df,\n",
    "        cmap=\"coolwarm\",\n",
    "        annot=True,\n",
    "        fmt=\".2f\",\n",
    "        cbar=False,\n",
    "        annot_kws={\"size\": 10},\n",
    "        mask=log2_fold_changes_df.isnull()  # Mask the missing values for better visualization\n",
    "    )\n",
    "\n",
    "    # Set the title\n",
    "    #plt.title(r'Log$_2$ Fold Changes', fontweight=\"bold\", fontsize=10)\n",
    "\n",
    "    # Turn off x-axis ticks\n",
    "    plt.xticks([])\n",
    "\n",
    "    # Set the font size for y-axis ticks\n",
    "    heatmap.yaxis.set_tick_params(labelsize=10)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    # Save the heatmap\n",
    "    plt.savefig('../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_log2foldchange_heatmap.png', dpi=1000, bbox_inches='tight')\n",
    "    print(\"The heatmap has been saved at '../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_log2foldchange_heatmap.png'\")\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "plot_lfd_heatmap(dea_results, gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the Base Mean heatmap for the genes of interest\n",
    "\n",
    "def plot_basemean_expression_heatmap(dea_results, gene_list):\n",
    "    \"\"\"\n",
    "    Plot the Base Mean heatmap for the Genes of Interest\n",
    "    \n",
    "    Parameters:\n",
    "        - dea_results (pandas.DataFrame): DESeq2 results with columns 'GeneSymbol', 'baseMean'.\n",
    "        - gene_list (list): List of gene symbols of interest.\n",
    "    \"\"\"\n",
    "\n",
    "    # Extract baseMean values for the genes of interest\n",
    "    goi_basemeans = [\n",
    "        dea_results.loc[dea_results[\"GeneSymbol\"] == gene, \"baseMean\"].values[0]\n",
    "        if not dea_results.loc[dea_results[\"GeneSymbol\"] == gene, \"baseMean\"].empty\n",
    "        else np.nan  # Use NaN instead of None for compatibility with heatmap\n",
    "        for gene in gene_list\n",
    "    ]\n",
    "\n",
    "    # Read oxidative damage response gene list\n",
    "    oxidative_damage_genes = gp.parser.read_gmt('../WP_OXIDATIVE_DAMAGE_RESPONSE.gmt')[\"WP_OXIDATIVE_DAMAGE_RESPONSE\"]\n",
    "\n",
    "    # Calculate the mean baseMean for oxidative damage response genes\n",
    "    oxidative_basemeans = [\n",
    "        dea_results.loc[dea_results[\"GeneSymbol\"] == gene, \"baseMean\"].values[0]\n",
    "        for gene in oxidative_damage_genes if gene in dea_results[\"GeneSymbol\"].values\n",
    "    ]\n",
    "    oxidative_damage_mean = np.nanmean(oxidative_basemeans)  # Avoid errors if list is empty\n",
    "\n",
    "    # Create a DataFrame including the mean baseMean of oxidative damage response genes\n",
    "    goi_basemeans_df = pd.DataFrame(\n",
    "        data=goi_basemeans + [oxidative_damage_mean],\n",
    "        index=gene_list + [\"WP_OXIDATIVE_\\nDAMAGE_RESPONSE\"],\n",
    "        columns=[\"baseMean\"]\n",
    "    )\n",
    "\n",
    "    # Plot heatmap\n",
    "    plt.figure(figsize=(0.75, 12.5))  # Adjust dynamically for better visualization\n",
    "    heatmap = sns.heatmap(\n",
    "        goi_basemeans_df,\n",
    "        cmap=\"viridis\",\n",
    "        annot=True,\n",
    "        fmt=\".2f\",\n",
    "        cbar=False,\n",
    "        annot_kws={\"size\": 10},\n",
    "        mask=goi_basemeans_df.isnull()  # Mask NaN values\n",
    "    )\n",
    "\n",
    "    # Set y-tick labels explicitly\n",
    "    heatmap.set_yticklabels(heatmap.get_yticklabels(), rotation=0, fontsize=10)\n",
    "\n",
    "    # Remove x-axis ticks\n",
    "    plt.xticks([])\n",
    "\n",
    "    plt.tight_layout()\n",
    "\n",
    "    # Save the figure with tight bounding box\n",
    "    save_path = '../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_basemean_heatmap.png'\n",
    "    plt.savefig(save_path, dpi=1000, bbox_inches='tight', pad_inches=0.1)\n",
    "    print(f\"The heatmap has been saved at '{save_path}'\")\n",
    "\n",
    "    # Display the plot\n",
    "    plt.show()\n",
    "\n",
    "plot_basemean_expression_heatmap(dea_results, gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plot_heatmap_with_clustermap(dds, gene_list, metadata):\n",
    "    \"\"\"\n",
    "    Plot a heatmap using the genes of interest for the different replicates.\n",
    "\n",
    "    Args:\n",
    "        dds (object): The DESeq2 object containing normalized counts and log1p transformed data.\n",
    "        gene_list (list): The list of genes to be used in constructing the heatmap.\n",
    "        metadata (DataFrame): The DataFrame containing metadata information of the replicates.\n",
    "\n",
    "    \"\"\"\n",
    "    # Extract the normalized counts and log1p transform them with a pseudocount of 1\n",
    "    #dds.layers['log1p'] = np.log1p(dds.layers['normed_counts'])\n",
    "    \n",
    "    # Extract the normalized counts of the genes of interest\n",
    "    dds_goi = dds[:, dds.var_names.isin(gene_list)]\n",
    "\n",
    "    # Create a DataFrame for plotting the heatmap\n",
    "    goi_norm_counts_df = pd.DataFrame(\n",
    "        dds_goi.layers['normed_counts'].T, \n",
    "        index=dds_goi.var_names, \n",
    "        columns=dds_goi.obs_names\n",
    "    )\n",
    "\n",
    "    # Extract the oxidative damage response genes from the WP_OXIDATIVE_DAMAGE_RESPONSE gene set\n",
    "    oxidative_damage_genes = gp.parser.read_gmt('../WP_OXIDATIVE_DAMAGE_RESPONSE.gmt')[\"WP_OXIDATIVE_DAMAGE_RESPONSE\"]\n",
    "\n",
    "    # Subset the AnnData object to include only the oxidative damage response genes\n",
    "    oxidative_damage_response_genes_dds = dds[:, dds.var_names.isin(oxidative_damage_genes)]\n",
    "\n",
    "    # Create a DataFrame of the normalized counts for the oxidative damage response genes\n",
    "    oxidative_damage_genes_norm_counts_df = pd.DataFrame(\n",
    "        oxidative_damage_response_genes_dds.layers['normed_counts'].T, \n",
    "        index=oxidative_damage_response_genes_dds.var_names, \n",
    "        columns=oxidative_damage_response_genes_dds.obs_names\n",
    "    )\n",
    "\n",
    "    # Compute the mean normalized counts for the oxidative damage response genes\n",
    "    oxidative_damage_genes_mean_norm_counts = oxidative_damage_genes_norm_counts_df.mean().tolist()\n",
    "\n",
    "    # Add the mean normalized counts to the goi_norm_counts_df\n",
    "    oxidative_damage_genes_mean_norm_counts_df=pd.DataFrame(data=[oxidative_damage_genes_mean_norm_counts], index=[\"WP_OXIDATIVE_\\nDAMAGE_RESPONSE\"], columns=oxidative_damage_response_genes_dds.obs_names)\n",
    "\n",
    "    goi_norm_counts_df=pd.concat([goi_norm_counts_df, oxidative_damage_genes_mean_norm_counts_df])\n",
    "\n",
    "    # Log1p transform the data\n",
    "    goi_norm_counts_df = np.log1p(goi_norm_counts_df)\n",
    "\n",
    "    # Get the indexes of control and covid conditions\n",
    "    indexes_condition_control = metadata[metadata['group'] == \"Control\"].index.tolist()\n",
    "    indexes_condition_covid = metadata[metadata['group'] == \"COVID-19\"].index.tolist()\n",
    "    indexes_order = indexes_condition_control + indexes_condition_covid\n",
    "\n",
    "\n",
    "    # Clean the data and reorder it based on gene_list and indexes_order\n",
    "    cleaned_data = goi_norm_counts_df[indexes_order]\n",
    "    cleaned_data = cleaned_data.reindex(gene_list+[\"WP_OXIDATIVE_\\nDAMAGE_RESPONSE\"])\n",
    "    cleaned_data.fillna(0, inplace=True)\n",
    "    cleaned_data = cleaned_data[sorted(cleaned_data.columns)]\n",
    "\n",
    "    \n",
    "    # Create a DataFrame to use as col_colors\n",
    "    col_colors_df = pd.DataFrame(index=cleaned_data.columns)\n",
    "\n",
    "    # highlight the control and the covid samples with different colors\n",
    "    col_colors_df.loc[indexes_condition_control, 'color'] = 'blue'  # Color for control columns\n",
    "    col_colors_df.loc[indexes_condition_covid, 'color'] = 'red'  # Color for covid columns\n",
    "\n",
    "    cleaned_data.columns = cleaned_data.columns.str[:-4]\n",
    "\n",
    "    # Generate heatmap with marked columns\n",
    "    heatmap = sns.clustermap(cleaned_data, method='average', z_score=-1, cmap='RdYlBu_r', col_cluster=False, row_cluster=False, col_colors=[col_colors_df['color']],  figsize=(18, 8), cbar_pos=(0.1, .6, 0.03, 0.18) )\n",
    "    heatmap.ax_heatmap.set_xticklabels(heatmap.ax_heatmap.get_xticklabels(), fontsize=12)\n",
    "    heatmap.ax_heatmap.set_yticklabels(heatmap.ax_heatmap.get_yticklabels(), fontsize=12)\n",
    "    heatmap.ax_heatmap.set_ylabel(\"\")\n",
    "\n",
    "\n",
    "    # Create the legend \n",
    "    legend = [Patch(color=color, label=label) for label, color in {'Control': 'blue', 'COVID-19': 'red'}.items()]\n",
    "\n",
    "    # Add legend to the plot\n",
    "    plt.legend(handles=legend, loc='upper left', bbox_to_anchor=(-.1, -0.2), frameon=False)\n",
    "\n",
    "    # Save the plot\n",
    "    plt.savefig('../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_normalized_expression_heatmap.png', dpi=1000)\n",
    "    print(\"Plot saved at '../Results/sc_transcriptomic_analysis/sc_transcriptomic_analysis_normalized_expression_heatmap.png'\")\n",
    "    # Show the plot\n",
    "    plt.show()\n",
    "    return cleaned_data\n",
    "\n",
    "# Generate the metadata for the samples and their associated condition\n",
    "metadata = pseudobulk_data.obs[[ \"group\"]]\n",
    "\n",
    "# Call the function to plot the heatmap\n",
    "plot_heatmap_with_clustermap(dds, gene_list, metadata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mutatioanalysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
